{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "# Un scénario complet"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Quelques informations et paramètres pour commencer "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "from __future__ import print_function\n",
    "\n",
    "def is_interactive():\n",
    "    import __main__ as main\n",
    "    return not hasattr(main, '__file__')\n",
    "\n",
    "if is_interactive():\n",
    "    Param_SBMLfile = \"../data/iLiverCancer1715.xml\"\n",
    "    Param_Pourcentage = 1.1\n",
    "    Param_RNASeqFilename = \"../data/LiverCancerTCGA_male.csv\"\n",
    "    Param_MaxIndegreeInRRG = 18\n",
    "    Param_Drug = \"Sorafenib (USAN/INN)\"\n",
    "    Param_NumberOfSample = 100\n",
    "    debugon = True\n",
    "    Param_DoReconstruction = False\n",
    "else:\n",
    "    import sys,getopt\n",
    "\n",
    "    optlist, args = getopt.getopt(sys.argv[1:], 'i:p:r:m:d:s:vR', [\"input\", \"percentage\",\"rna\",\"maxdegree\",\"drug\",\"samples\",\"verbose\",\"with-reconstruction\"])\n",
    "\n",
    "    for o,v in optlist:\n",
    "        if  o in (\"-i\",\"--input\"):\n",
    "            Param_SBMLfile = v\n",
    "        else:\n",
    "            Param_SBMLfile = \"../data/iLiverCancer1715.xml\"\n",
    "    \n",
    "        if o in (\"-r\",\"--rna\"):\n",
    "            Param_RNASeqFilename = v\n",
    "        else:\n",
    "            Param_RNASeqFilename = \"../data/LiverCancerTCGA_male.csv\"\n",
    "    \n",
    "        if o in (\"-p\",\"--percentage\"):\n",
    "            Param_Pourcentage = float(v)\n",
    "        else:\n",
    "            Param_Pourcentage = 1.1\n",
    "    \n",
    "        if o in (\"-m\",\"--maxdegree\"):\n",
    "            Param_MaxIndegreeInRRG = int(v)\n",
    "        else:\n",
    "            Param_MaxIndegreeInRRG = 18\n",
    "    \n",
    "        if o in (\"-d\",\"--drug\"):\n",
    "            Param_Drug = v\n",
    "        else:\n",
    "            Param_Drug = \"Sorafenib (USAN/INN)\"\n",
    "    \n",
    "        if o in (\"-s\",\"--samples\"):\n",
    "            Param_NumberOfSample = int(v)\n",
    "        else:\n",
    "            Param_NumberOfSample = 100\n",
    "    \n",
    "        if o in (\"-v\",\"--verbose\"):\n",
    "            debugon = True\n",
    "        else:\n",
    "            debugon = False\n",
    "            \n",
    "        if o in (\"-R\",\"--with-reconstruction\"):\n",
    "            Param_DoReconstruction = True\n",
    "        else:\n",
    "            Param_DoReconstruction = False\n",
    "\n",
    "# internal variables\n",
    "#import shutils\n",
    "#shutils.copy(Param_SBMLfile,\"../tmp\")\n",
    "import os\n",
    "\n",
    "Param_SBMLfileBN = \"../tmp/\"+os.path.basename(Param_SBMLfile)\n",
    "\n",
    "PC_Endpoint = \"http://rdf.pathwaycommons.org/sparql\"\n",
    "RXN_EC_file = Param_SBMLfileBN.replace(\".xml\",\".rxn\")\n",
    "BIF_file = Param_SBMLfileBN.replace(\".xml\",\".bif\")\n",
    "Param_RRGFilename = Param_SBMLfileBN.replace(\".xml\",\".rrg\")\n",
    "Param_ProbaReactionsFilename = Param_SBMLfileBN.replace(\".xml\",\".prob\")\n",
    "#whichsolver = \"gurobi\"\n",
    "whichsolver = \"cglpk\"\n",
    "\n",
    "\n",
    "filenameCorrespondanceEnsemblGenenames = \"../tables/Correspondance_Ensembl_genename.txt\"\n",
    "filenameCorrespondanceEnsemblEnzymes = \"../tables/Correspondance_Enzymes_genenames.txt\"\n",
    "filenameCorrespondanceEnsemblTransporters = \"../tables/Correspondance_Transporters_genenames.txt\"\n",
    "filenameKEGGDrugsGenes = \"../tables/Public_KEGGDRUG_Targets.csv\"\n",
    "filenameCorrespondanceEnsemblMeanLength = \"../tables/Correspondance_Ensembl_MeanTranscriptSize.txt\"\n",
    "\n",
    "\n",
    "# internal function can be edited\n",
    "def Binarized(x,m,s,gene=\"null\"):\n",
    "    x=log(0.0001+x)\n",
    "    InfBoundCoeff=-0.46 #0.46\n",
    "    SupBoundCoeff=10000000 # 0.46\n",
    "    if (x<m-InfBoundCoeff*s):\n",
    "        return 0\n",
    "    if (x>m+SupBoundCoeff*s):\n",
    "        return 2\n",
    "    return 1\n",
    "\n",
    "\n",
    "def Binarized2(x,m,s,gene=\"null\"):\n",
    "    InfBoundCoeff=0.1\n",
    "    m=AllGeneMeans.get(gene)\n",
    "    if (m):\n",
    "        if (m !=0):\n",
    "            if (x/m<InfBoundCoeff):\n",
    "                return 0\n",
    "            else:\n",
    "                return 1\n",
    "    return 0\n",
    "\n",
    "def Binarized3(x,m,s,gene=\"null\"):\n",
    "    InfBoundCoeff=1\n",
    "#    m=AllGeneMeans.get(gene)\n",
    "    m=1\n",
    "    if (m):\n",
    "        if (m !=0):\n",
    "            if (x/m<InfBoundCoeff):\n",
    "                return 0\n",
    "            else:\n",
    "                return 1\n",
    "    return 0"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Tous les includes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import optlang # NB : due to a bug in cobrapy, optlang must be imported before cobra\n",
    "import cobra\n",
    "from cobra.flux_analysis import flux_variability_analysis\n",
    "\n",
    "from timeit import default_timer as timer\n",
    "import csv\n",
    "from SPARQLWrapper import SPARQLWrapper, JSON\n",
    "import networkx as nx\n",
    "from math import log\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "from pgmpy.models import BayesianModel\n",
    "from pgmpy.estimators import MaximumLikelihoodEstimator, BayesianEstimator\n",
    "from pgmpy.factors.discrete import State\n",
    "from pgmpy.sampling import BayesianModelSampling\n",
    "from pgmpy.readwrite import BIFReader, BIFWriter\n",
    "\n",
    "\n",
    "#from optlang import Model,Variable,Constraint,Objective\n",
    "\n",
    "import sys\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# a few definition to print on stderr\n",
    "def eprint(*args, **kwargs):\n",
    "    if debugon:\n",
    "        print(*args, file=sys.stderr, **kwargs)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Testing ../data/iLiverCancer1715.xml for Sorafenib (USAN/INN)\n"
     ]
    }
   ],
   "source": [
    "## DEBUG ONLY\n",
    "begintotal = timer()\n",
    "eprint(\"Testing\",Param_SBMLfile,\"for\",Param_Drug)\n",
    "\n",
    "## END DEBUG ONLY"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Chargement des listes de correspondances"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Importing all the correspondance tables\n",
      "Missing ENSG00000273504\n",
      "Missing ENSG00000273527\n",
      "Missing ENSG00000273530\n",
      "Missing ENSG00000273531\n",
      "Missing ENSG00000273534\n",
      "Missing ENSG00000273545\n",
      "Missing ENSG00000273556\n",
      "Missing ENSG00000273560\n",
      "Missing ENSG00000273561\n",
      "Missing ENSG00000273577\n",
      "Missing ENSG00000273581\n",
      "Missing ENSG00000273583\n",
      "Missing ENSG00000273615\n",
      "Missing ENSG00000273626\n",
      "Missing ENSG00000273630\n",
      "Missing ENSG00000273631\n",
      "Missing ENSG00000273632\n",
      "Missing ENSG00000273653\n",
      "Missing ENSG00000273662\n",
      "Missing ENSG00000273663\n",
      "Missing ENSG00000273665\n",
      "Missing ENSG00000273667\n",
      "Missing ENSG00000273672\n",
      "Missing ENSG00000273678\n",
      "Missing ENSG00000273684\n",
      "Missing ENSG00000273685\n",
      "Missing ENSG00000273689\n",
      "Missing ENSG00000273690\n",
      "Missing ENSG00000273695\n",
      "Missing ENSG00000273697\n",
      "Missing ENSG00000273699\n",
      "Missing ENSG00000273705\n",
      "Missing ENSG00000273718\n",
      "Missing ENSG00000273743\n",
      "Missing ENSG00000273751\n",
      "Missing ENSG00000273754\n",
      "Missing ENSG00000273774\n",
      "Missing ENSG00000273781\n",
      "Missing ENSG00000273787\n",
      "Missing ENSG00000273789\n",
      "Missing ENSG00000273790\n",
      "Missing ENSG00000273798\n",
      "Missing ENSG00000273801\n",
      "Missing ENSG00000273809\n",
      "Missing ENSG00000273810\n",
      "Missing ENSG00000273811\n",
      "Missing ENSG00000273817\n",
      "Missing ENSG00000273823\n",
      "Missing ENSG00000273826\n",
      "Missing ENSG00000273832\n",
      "Missing ENSG00000273834\n",
      "Missing ENSG00000273842\n",
      "Missing ENSG00000273848\n",
      "Missing ENSG00000273854\n",
      "Missing ENSG00000273857\n",
      "Missing ENSG00000273862\n",
      "Missing ENSG00000273868\n",
      "Missing ENSG00000273878\n",
      "Missing ENSG00000273879\n",
      "Missing ENSG00000273892\n",
      "Missing ENSG00000273895\n",
      "Missing ENSG00000273902\n",
      "Missing ENSG00000273910\n",
      "Missing ENSG00000273915\n",
      "Missing ENSG00000273917\n",
      "Missing ENSG00000273926\n",
      "Missing ENSG00000273930\n",
      "Missing ENSG00000273939\n",
      "Missing ENSG00000273941\n",
      "Missing ENSG00000273949\n",
      "Missing ENSG00000273952\n",
      "Missing ENSG00000273960\n",
      "Missing ENSG00000273985\n",
      "Missing ENSG00000273996\n",
      "Missing ENSG00000274013\n",
      "Missing ENSG00000274014\n",
      "Missing ENSG00000274027\n",
      "Missing ENSG00000274037\n",
      "Missing ENSG00000274043\n",
      "Missing ENSG00000274057\n",
      "Missing ENSG00000274061\n",
      "Missing ENSG00000274065\n",
      "Missing ENSG00000274074\n",
      "Missing ENSG00000274077\n",
      "Missing ENSG00000274083\n",
      "Missing ENSG00000274087\n",
      "Missing ENSG00000274088\n",
      "Missing ENSG00000274094\n",
      "Missing ENSG00000274095\n",
      "Missing ENSG00000274096\n",
      "Missing ENSG00000274103\n",
      "Missing ENSG00000274140\n",
      "Missing ENSG00000274141\n",
      "Missing ENSG00000274151\n",
      "Missing ENSG00000274153\n",
      "Missing ENSG00000274155\n",
      "Missing ENSG00000274160\n",
      "Missing ENSG00000274182\n",
      "Missing ENSG00000274188\n",
      "Missing ENSG00000274190\n",
      "Missing ENSG00000274200\n",
      "Missing ENSG00000274203\n",
      "Missing ENSG00000274207\n",
      "Missing ENSG00000274222\n",
      "Missing ENSG00000274223\n",
      "Missing ENSG00000274237\n",
      "Missing ENSG00000274241\n",
      "Missing ENSG00000274246\n",
      "Missing ENSG00000274250\n",
      "Missing ENSG00000274255\n",
      "Missing ENSG00000274268\n",
      "Missing ENSG00000274277\n",
      "Missing ENSG00000274278\n",
      "Missing ENSG00000274289\n",
      "Missing ENSG00000274298\n",
      "Missing ENSG00000274301\n",
      "Missing ENSG00000274302\n",
      "Missing ENSG00000274304\n",
      "Missing ENSG00000274325\n",
      "Missing ENSG00000274329\n",
      "Missing ENSG00000274335\n",
      "Missing ENSG00000274358\n",
      "Missing ENSG00000274366\n",
      "Missing ENSG00000274368\n",
      "Missing ENSG00000274377\n",
      "Missing ENSG00000274388\n",
      "Missing ENSG00000274389\n",
      "Missing ENSG00000274416\n",
      "Missing ENSG00000274426\n",
      "Missing ENSG00000274431\n",
      "Missing ENSG00000274433\n",
      "Missing ENSG00000274448\n",
      "Missing ENSG00000274453\n",
      "Missing ENSG00000274464\n",
      "Missing ENSG00000274470\n",
      "Missing ENSG00000274479\n",
      "Missing ENSG00000274483\n",
      "Missing ENSG00000274490\n",
      "Missing ENSG00000274496\n",
      "Missing ENSG00000274509\n",
      "Missing ENSG00000274519\n",
      "Missing ENSG00000274521\n",
      "Missing ENSG00000274524\n",
      "Missing ENSG00000274526\n",
      "Missing ENSG00000274537\n",
      "Missing ENSG00000274548\n",
      "Missing ENSG00000274550\n",
      "Missing ENSG00000274555\n",
      "Missing ENSG00000274557\n",
      "Missing ENSG00000274590\n",
      "Missing ENSG00000274592\n",
      "Missing ENSG00000274596\n",
      "Missing ENSG00000274604\n",
      "Missing ENSG00000274622\n",
      "Missing ENSG00000274633\n",
      "Missing ENSG00000274634\n",
      "Missing ENSG00000274635\n",
      "Missing ENSG00000274645\n",
      "Missing ENSG00000274648\n",
      "Missing ENSG00000274668\n",
      "Missing ENSG00000274673\n",
      "Missing ENSG00000274674\n",
      "Missing ENSG00000274683\n",
      "Missing ENSG00000274689\n",
      "Missing ENSG00000274694\n",
      "Missing ENSG00000274725\n",
      "Missing ENSG00000274728\n",
      "Missing ENSG00000274735\n",
      "Missing ENSG00000274743\n",
      "Missing ENSG00000274753\n",
      "Missing ENSG00000274761\n",
      "Missing ENSG00000274773\n",
      "Missing ENSG00000274778\n",
      "Missing ENSG00000274784\n",
      "Missing ENSG00000274787\n",
      "Missing ENSG00000274793\n",
      "Missing ENSG00000274794\n",
      "Missing ENSG00000274802\n",
      "Missing ENSG00000274815\n",
      "Missing ENSG00000274831\n",
      "Missing ENSG00000274856\n",
      "Missing ENSG00000274871\n",
      "Missing ENSG00000274872\n",
      "Missing ENSG00000274877\n",
      "Missing ENSG00000274880\n",
      "Missing ENSG00000274888\n",
      "Missing ENSG00000274890\n",
      "Missing ENSG00000274900\n",
      "Missing ENSG00000274910\n",
      "Missing ENSG00000274913\n",
      "Missing ENSG00000274924\n",
      "Missing ENSG00000274951\n",
      "Missing ENSG00000274961\n",
      "Missing ENSG00000274982\n",
      "Missing ENSG00000274983\n",
      "Missing ENSG00000274991\n",
      "Missing ENSG00000274992\n",
      "Missing ENSG00000275000\n",
      "Missing ENSG00000275010\n",
      "Missing ENSG00000275020\n",
      "Missing ENSG00000275027\n",
      "Missing ENSG00000275039\n",
      "Missing ENSG00000275042\n",
      "Missing ENSG00000275044\n",
      "Missing ENSG00000275053\n",
      "Missing ENSG00000275058\n",
      "Missing ENSG00000275065\n",
      "Missing ENSG00000275076\n",
      "Missing ENSG00000275077\n",
      "Missing ENSG00000275085\n",
      "Missing ENSG00000275089\n",
      "Missing ENSG00000275096\n",
      "Missing ENSG00000275098\n",
      "Missing ENSG00000275100\n",
      "Missing ENSG00000275105\n",
      "Missing ENSG00000275114\n",
      "Missing ENSG00000275124\n",
      "Missing ENSG00000275135\n",
      "Missing ENSG00000275147\n",
      "Missing ENSG00000275149\n",
      "Missing ENSG00000275159\n",
      "Missing ENSG00000275182\n",
      "Missing ENSG00000275184\n",
      "Missing ENSG00000275200\n",
      "Missing ENSG00000275209\n",
      "Missing ENSG00000275218\n",
      "Missing ENSG00000275220\n",
      "Missing ENSG00000275224\n",
      "Missing ENSG00000275233\n",
      "Missing ENSG00000275235\n",
      "Missing ENSG00000275241\n",
      "Missing ENSG00000275251\n",
      "Missing ENSG00000275252\n",
      "Missing ENSG00000275258\n",
      "Missing ENSG00000275261\n",
      "Missing ENSG00000275264\n",
      "Missing ENSG00000275267\n",
      "Missing ENSG00000275270\n",
      "Missing ENSG00000275271\n",
      "Missing ENSG00000275277\n",
      "Missing ENSG00000275279\n",
      "Missing ENSG00000275292\n",
      "Missing ENSG00000275301\n",
      "Missing ENSG00000275303\n",
      "Missing ENSG00000275309\n",
      "Missing ENSG00000275364\n",
      "Missing ENSG00000275370\n",
      "Missing ENSG00000275378\n",
      "Missing ENSG00000275382\n",
      "Missing ENSG00000275398\n",
      "Missing ENSG00000275402\n",
      "Missing ENSG00000275403\n",
      "Missing ENSG00000275425\n",
      "Missing ENSG00000275439\n",
      "Missing ENSG00000275446\n",
      "Missing ENSG00000275447\n",
      "Missing ENSG00000275459\n",
      "Missing ENSG00000275466\n",
      "Missing ENSG00000275505\n",
      "Missing ENSG00000275507\n",
      "Missing ENSG00000275509\n",
      "Missing ENSG00000275510\n",
      "Missing ENSG00000275530\n",
      "Missing ENSG00000275533\n",
      "Missing ENSG00000275534\n",
      "Missing ENSG00000275535\n",
      "Missing ENSG00000275537\n",
      "Missing ENSG00000275541\n",
      "Missing ENSG00000275554\n",
      "Missing ENSG00000275595\n",
      "Missing ENSG00000275603\n",
      "Missing ENSG00000275604\n",
      "Missing ENSG00000275622\n",
      "Missing ENSG00000275653\n",
      "Missing ENSG00000275657\n",
      "Missing ENSG00000275668\n",
      "Missing ENSG00000275677\n",
      "Missing ENSG00000275682\n",
      "Missing ENSG00000275685\n",
      "Missing ENSG00000275686\n",
      "Missing ENSG00000275687\n",
      "Missing ENSG00000275690\n",
      "Missing ENSG00000275694\n",
      "Missing ENSG00000275701\n",
      "Missing ENSG00000275702\n",
      "Missing ENSG00000275704\n",
      "Missing ENSG00000275707\n",
      "Missing ENSG00000275727\n",
      "Missing ENSG00000275729\n",
      "Missing ENSG00000275739\n",
      "Missing ENSG00000275742\n",
      "Missing ENSG00000275761\n",
      "Missing ENSG00000275783\n",
      "Missing ENSG00000275790\n",
      "Missing ENSG00000275802\n",
      "Missing ENSG00000275813\n",
      "Missing ENSG00000275820\n",
      "Missing ENSG00000275829\n",
      "Missing ENSG00000275831\n",
      "Missing ENSG00000275845\n",
      "Missing ENSG00000275847\n",
      "Missing ENSG00000275849\n",
      "Missing ENSG00000275858\n",
      "Missing ENSG00000275864\n",
      "Missing ENSG00000275870\n",
      "Missing ENSG00000275876\n",
      "Missing ENSG00000275883\n",
      "Missing ENSG00000275884\n",
      "Missing ENSG00000275891\n",
      "Missing ENSG00000275892\n",
      "Missing ENSG00000275899\n",
      "Missing ENSG00000275912\n",
      "Missing ENSG00000275916\n",
      "Missing ENSG00000275922\n",
      "Missing ENSG00000275925\n",
      "Missing ENSG00000275926\n",
      "Missing ENSG00000275934\n",
      "Missing ENSG00000275943\n",
      "Missing ENSG00000275952\n",
      "Missing ENSG00000275955\n",
      "Missing ENSG00000275957\n",
      "Missing ENSG00000275977\n",
      "Missing ENSG00000275986\n",
      "Missing ENSG00000275998\n",
      "Missing ENSG00000276018\n",
      "Missing ENSG00000276024\n",
      "Missing ENSG00000276037\n",
      "Missing ENSG00000276038\n",
      "Missing ENSG00000276056\n",
      "Missing ENSG00000276059\n",
      "Missing ENSG00000276078\n",
      "Missing ENSG00000276079\n",
      "Missing ENSG00000276090\n",
      "Missing ENSG00000276104\n",
      "Missing ENSG00000276108\n",
      "Missing ENSG00000276117\n",
      "Missing ENSG00000276151\n",
      "Missing ENSG00000276157\n",
      "Missing ENSG00000276165\n",
      "Missing ENSG00000276171\n",
      "Missing ENSG00000276198\n",
      "Missing ENSG00000276201\n",
      "Missing ENSG00000276215\n",
      "Missing ENSG00000276235\n",
      "Missing ENSG00000276236\n",
      "Missing ENSG00000276237\n",
      "Missing ENSG00000276242\n",
      "Missing ENSG00000276244\n",
      "Missing ENSG00000276246\n",
      "Missing ENSG00000276252\n",
      "Missing ENSG00000276271\n",
      "Missing ENSG00000276273\n",
      "Missing ENSG00000276280\n",
      "Missing ENSG00000276283\n",
      "Missing ENSG00000276288\n",
      "Missing ENSG00000276301\n",
      "Missing ENSG00000276305\n",
      "Missing ENSG00000276315\n",
      "Missing ENSG00000276322\n",
      "Missing ENSG00000276325\n",
      "Missing ENSG00000276331\n",
      "Missing ENSG00000276335\n",
      "Missing ENSG00000276338\n",
      "Missing ENSG00000276347\n",
      "Missing ENSG00000276354\n",
      "Missing ENSG00000276369\n",
      "Missing ENSG00000276372\n",
      "Missing ENSG00000276373\n",
      "Missing ENSG00000276381\n",
      "Missing ENSG00000276389\n",
      "Missing ENSG00000276401\n",
      "Missing ENSG00000276419\n",
      "Missing ENSG00000276444\n",
      "Missing ENSG00000276456\n",
      "Missing ENSG00000276472\n",
      "Missing ENSG00000276477\n",
      "Missing ENSG00000276499\n",
      "Missing ENSG00000276506\n",
      "Missing ENSG00000276510\n",
      "Missing ENSG00000276511\n",
      "Missing ENSG00000276513\n",
      "Missing ENSG00000276515\n",
      "Missing ENSG00000276543\n",
      "Missing ENSG00000276554\n",
      "Missing ENSG00000276555\n",
      "Missing ENSG00000276558\n",
      "Missing ENSG00000276574\n",
      "Missing ENSG00000276587\n",
      "Missing ENSG00000276602\n",
      "Missing ENSG00000276607\n",
      "Missing ENSG00000276614\n",
      "Missing ENSG00000276617\n",
      "Missing ENSG00000276619\n",
      "Missing ENSG00000276621\n",
      "Missing ENSG00000276624\n",
      "Missing ENSG00000276629\n",
      "Missing ENSG00000276635\n",
      "Missing ENSG00000276648\n",
      "Missing ENSG00000276652\n",
      "Missing ENSG00000276659\n",
      "Missing ENSG00000276660\n",
      "Missing ENSG00000276661\n",
      "Missing ENSG00000276677\n",
      "Missing ENSG00000276682\n",
      "Missing ENSG00000276683\n",
      "Missing ENSG00000276694\n",
      "Missing ENSG00000276695\n",
      "Missing ENSG00000276737\n",
      "Missing ENSG00000276741\n",
      "Missing ENSG00000276743\n",
      "Missing ENSG00000276745\n",
      "Missing ENSG00000276748\n",
      "Missing ENSG00000276750\n",
      "Missing ENSG00000276751\n",
      "Missing ENSG00000276752\n",
      "Missing ENSG00000276762\n",
      "Missing ENSG00000276773\n",
      "Missing ENSG00000276783\n",
      "Missing ENSG00000276792\n",
      "Missing ENSG00000276793\n",
      "Missing ENSG00000276796\n",
      "Missing ENSG00000276797\n",
      "Missing ENSG00000276801\n",
      "Missing ENSG00000276811\n",
      "Missing ENSG00000276812\n",
      "Missing ENSG00000276813\n",
      "Missing ENSG00000276815\n",
      "Missing ENSG00000276817\n",
      "Missing ENSG00000276818\n",
      "Missing ENSG00000276827\n",
      "Missing ENSG00000276852\n",
      "Missing ENSG00000276862\n",
      "Missing ENSG00000276877\n",
      "Missing ENSG00000276883\n",
      "Missing ENSG00000276889\n",
      "Missing ENSG00000276893\n",
      "Missing ENSG00000276899\n",
      "Missing ENSG00000276921\n",
      "Missing ENSG00000276922\n",
      "Missing ENSG00000276924\n",
      "Missing ENSG00000276927\n",
      "Missing ENSG00000276933\n",
      "Missing ENSG00000276939\n",
      "Missing ENSG00000276942\n",
      "Missing ENSG00000276943\n",
      "Missing ENSG00000276962\n",
      "Missing ENSG00000276967\n",
      "Missing ENSG00000276978\n",
      "Missing ENSG00000276987\n",
      "Missing ENSG00000276989\n",
      "Missing ENSG00000276990\n",
      "Missing ENSG00000276995\n",
      "Missing ENSG00000277002\n",
      "Missing ENSG00000277021\n",
      "Missing ENSG00000277023\n",
      "Missing ENSG00000277026\n",
      "Missing ENSG00000277045\n",
      "Missing ENSG00000277054\n",
      "Missing ENSG00000277061\n",
      "Missing ENSG00000277064\n",
      "Missing ENSG00000277074\n",
      "Missing ENSG00000277082\n",
      "Missing ENSG00000277097\n",
      "Missing ENSG00000277100\n",
      "Missing ENSG00000277102\n",
      "Missing ENSG00000277105\n",
      "Missing ENSG00000277115\n",
      "Missing ENSG00000277116\n",
      "Missing ENSG00000277121\n",
      "Missing ENSG00000277122\n",
      "Missing ENSG00000277123\n",
      "Missing ENSG00000277137\n",
      "Missing ENSG00000277155\n",
      "Missing ENSG00000277158\n",
      "Missing ENSG00000277160\n",
      "Missing ENSG00000277169\n",
      "Missing ENSG00000277180\n",
      "Missing ENSG00000277189\n",
      "Missing ENSG00000277197\n",
      "Missing ENSG00000277201\n",
      "Missing ENSG00000277204\n",
      "Missing ENSG00000277207\n",
      "Missing ENSG00000277212\n",
      "Missing ENSG00000277216\n",
      "Missing ENSG00000277225\n",
      "Missing ENSG00000277231\n",
      "Missing ENSG00000277239\n",
      "Missing ENSG00000277244\n",
      "Missing ENSG00000277257\n",
      "Missing ENSG00000277259\n",
      "Missing ENSG00000277275\n",
      "Missing ENSG00000277284\n",
      "Missing ENSG00000277285\n",
      "Missing ENSG00000277294\n",
      "Missing ENSG00000277296\n",
      "Missing ENSG00000277300\n",
      "Missing ENSG00000277302\n",
      "Missing ENSG00000277305\n",
      "Missing ENSG00000277312\n",
      "Missing ENSG00000277321\n",
      "Missing ENSG00000277323\n",
      "Missing ENSG00000277325\n",
      "Missing ENSG00000277326\n",
      "Missing ENSG00000277328\n",
      "Missing ENSG00000277329\n",
      "Missing ENSG00000277330\n",
      "Missing ENSG00000277331\n",
      "Missing ENSG00000277333\n",
      "Missing ENSG00000277338\n",
      "Missing ENSG00000277346\n",
      "Missing ENSG00000277347\n",
      "Missing ENSG00000277348\n",
      "Missing ENSG00000277353\n",
      "Missing ENSG00000277360\n",
      "Missing ENSG00000277372\n",
      "Missing ENSG00000277393\n",
      "Missing ENSG00000277394\n",
      "Missing ENSG00000277433\n",
      "Missing ENSG00000277441\n",
      "Missing ENSG00000277451\n",
      "Missing ENSG00000277472\n",
      "Missing ENSG00000277487\n",
      "Missing ENSG00000277489\n",
      "Missing ENSG00000277497\n",
      "Missing ENSG00000277503\n",
      "Missing ENSG00000277520\n",
      "Missing ENSG00000277528\n",
      "Missing ENSG00000277537\n",
      "Missing ENSG00000277547\n",
      "Missing ENSG00000277555\n",
      "Missing ENSG00000277573\n",
      "Missing ENSG00000277593\n",
      "Missing ENSG00000277601\n",
      "Missing ENSG00000277609\n",
      "Missing ENSG00000277612\n",
      "Missing ENSG00000277622\n",
      "Missing ENSG00000277627\n",
      "Missing ENSG00000277644\n",
      "Missing ENSG00000277645\n",
      "Missing ENSG00000277648\n",
      "Missing ENSG00000277652\n",
      "Missing ENSG00000277663\n",
      "Missing ENSG00000277664\n",
      "Missing ENSG00000277665\n",
      "Missing ENSG00000277682\n",
      "Missing ENSG00000277683\n",
      "Missing ENSG00000277685\n",
      "Missing ENSG00000277699\n",
      "Missing ENSG00000277700\n",
      "Missing ENSG00000277703\n",
      "Missing ENSG00000277712\n",
      "Missing ENSG00000277714\n",
      "Missing ENSG00000277716\n",
      "Missing ENSG00000277719\n",
      "Missing ENSG00000277724\n",
      "Missing ENSG00000277727\n",
      "Missing ENSG00000277731\n",
      "Missing ENSG00000277735\n",
      "Missing ENSG00000277736\n",
      "Missing ENSG00000277756\n",
      "Missing ENSG00000277760\n",
      "Missing ENSG00000277772\n",
      "Missing ENSG00000277779\n",
      "Missing ENSG00000277780\n",
      "Missing ENSG00000277787\n",
      "Missing ENSG00000277792\n",
      "Missing ENSG00000277793\n",
      "Missing ENSG00000277798\n",
      "Missing ENSG00000277805\n",
      "Missing ENSG00000277810\n",
      "Missing ENSG00000277812\n",
      "Missing ENSG00000277813\n",
      "Missing ENSG00000277814\n",
      "Missing ENSG00000277815\n",
      "Missing ENSG00000277819\n",
      "Missing ENSG00000277821\n",
      "Missing ENSG00000277822\n",
      "Missing ENSG00000277827\n",
      "Missing ENSG00000277832\n",
      "Missing ENSG00000277838\n",
      "Missing ENSG00000277839\n",
      "Missing ENSG00000277843\n",
      "Missing ENSG00000277847\n",
      "Missing ENSG00000277854\n",
      "Missing ENSG00000277861\n",
      "Missing ENSG00000277875\n",
      "Missing ENSG00000277877\n",
      "Missing ENSG00000277884\n",
      "Missing ENSG00000277886\n",
      "Missing ENSG00000277896\n",
      "Missing ENSG00000277910\n",
      "Missing ENSG00000277929\n",
      "Missing ENSG00000277933\n",
      "Missing ENSG00000277934\n",
      "Missing ENSG00000277937\n",
      "Missing ENSG00000277939\n",
      "Missing ENSG00000277951\n",
      "Missing ENSG00000277960\n",
      "Missing ENSG00000277962\n",
      "Missing ENSG00000277990\n",
      "Missing ENSG00000277992\n",
      "Missing ENSG00000278005\n",
      "Missing ENSG00000278014\n",
      "Missing ENSG00000278027\n",
      "Missing ENSG00000278047\n",
      "Missing ENSG00000278056\n",
      "Missing ENSG00000278059\n",
      "Missing ENSG00000278065\n",
      "Missing ENSG00000278067\n",
      "Missing ENSG00000278076\n",
      "Missing ENSG00000278093\n",
      "Missing ENSG00000278105\n",
      "Missing ENSG00000278125\n",
      "Missing ENSG00000278136\n",
      "Missing ENSG00000278142\n",
      "Missing ENSG00000278148\n",
      "Missing ENSG00000278150\n",
      "Missing ENSG00000278166\n",
      "Missing ENSG00000278176\n",
      "Missing ENSG00000278179\n",
      "Missing ENSG00000278185\n",
      "Missing ENSG00000278187\n",
      "Missing ENSG00000278203\n",
      "Missing ENSG00000278219\n",
      "Missing ENSG00000278227\n",
      "Missing ENSG00000278228\n",
      "Missing ENSG00000278244\n",
      "Missing ENSG00000278251\n",
      "Missing ENSG00000278258\n",
      "Missing ENSG00000278260\n",
      "Missing ENSG00000278282\n",
      "Missing ENSG00000278285\n",
      "Missing ENSG00000278290\n",
      "Missing ENSG00000278297\n",
      "Missing ENSG00000278309\n",
      "Missing ENSG00000278321\n",
      "Missing ENSG00000278335\n",
      "Missing ENSG00000278350\n",
      "Missing ENSG00000278352\n",
      "Missing ENSG00000278353\n",
      "Missing ENSG00000278366\n",
      "Missing ENSG00000278371\n",
      "Missing ENSG00000278375\n",
      "Missing ENSG00000278392\n",
      "Missing ENSG00000278393\n",
      "Missing ENSG00000278398\n",
      "Missing ENSG00000278405\n",
      "Missing ENSG00000278423\n",
      "Missing ENSG00000278435\n",
      "Missing ENSG00000278439\n",
      "Missing ENSG00000278452\n",
      "Missing ENSG00000278453\n",
      "Missing ENSG00000278460\n",
      "Missing ENSG00000278471\n",
      "Missing ENSG00000278479\n",
      "Missing ENSG00000278491\n",
      "Missing ENSG00000278504\n",
      "Missing ENSG00000278506\n",
      "Missing ENSG00000278507\n",
      "Missing ENSG00000278509\n",
      "Missing ENSG00000278512\n",
      "Missing ENSG00000278521\n",
      "Missing ENSG00000278534\n",
      "Missing ENSG00000278536\n",
      "Missing ENSG00000278544\n",
      "Missing ENSG00000278547\n",
      "Missing ENSG00000278556\n",
      "Missing ENSG00000278557\n",
      "Missing ENSG00000278560\n",
      "Missing ENSG00000278564\n",
      "Missing ENSG00000278568\n",
      "Missing ENSG00000278575\n",
      "Missing ENSG00000278578\n",
      "Missing ENSG00000278579\n",
      "Missing ENSG00000278584\n",
      "Missing ENSG00000278597\n",
      "Missing ENSG00000278609\n",
      "Missing ENSG00000278612\n",
      "Missing ENSG00000278628\n",
      "Missing ENSG00000278629\n",
      "Missing ENSG00000278639\n",
      "Missing ENSG00000278645\n",
      "Missing ENSG00000278648\n",
      "Missing ENSG00000278651\n",
      "Missing ENSG00000278652\n",
      "Missing ENSG00000278660\n",
      "Missing ENSG00000278663\n",
      "Missing ENSG00000278669\n",
      "Missing ENSG00000278671\n",
      "Missing ENSG00000278675\n",
      "Missing ENSG00000278679\n",
      "Missing ENSG00000278686\n",
      "Missing ENSG00000278717\n",
      "Missing ENSG00000278748\n",
      "Missing ENSG00000278750\n",
      "Missing ENSG00000278751\n",
      "Missing ENSG00000278753\n",
      "Missing ENSG00000278772\n",
      "Missing ENSG00000278792\n",
      "Missing ENSG00000278804\n",
      "Missing ENSG00000278810\n",
      "Missing ENSG00000278814\n",
      "Missing ENSG00000278820\n",
      "Missing ENSG00000278832\n",
      "Missing ENSG00000278841\n",
      "Missing ENSG00000278844\n",
      "Missing ENSG00000278882\n",
      "Missing ENSG00000278947\n",
      "Missing ENSG00000279003\n",
      "Missing ENSG00000279027\n",
      "Missing ENSG00000279037\n",
      "Missing ENSG00000279038\n",
      "Missing ENSG00000279185\n",
      "Missing ENSG00000279188\n",
      "Missing ENSG00000279210\n",
      "Missing ENSG00000279216\n",
      "Missing ENSG00000279242\n",
      "Missing ENSG00000279252\n",
      "Missing ENSG00000279260\n",
      "Missing ENSG00000279272\n",
      "Missing ENSG00000279291\n",
      "Missing ENSG00000279338\n",
      "Missing ENSG00000279424\n",
      "Missing ENSG00000279468\n",
      "Missing ENSG00000279496\n",
      "Missing ENSG00000279556\n",
      "Missing ENSG00000279582\n",
      "Missing ENSG00000279618\n",
      "Missing ENSG00000279624\n",
      "Missing ENSG00000279723\n",
      "Missing ENSG00000279802\n",
      "Missing ENSG00000279831\n",
      "Missing ENSG00000279843\n",
      "Missing ENSG00000279866\n",
      "Missing ENSG00000279868\n",
      "Missing ENSG00000279888\n",
      "Missing ENSG00000279893\n",
      "Missing ENSG00000279911\n",
      "Missing ENSG00000279922\n",
      "Missing ENSG00000279941\n",
      "Missing ENSG00000279975\n",
      "Missing ENSG00000279978\n",
      "Missing ENSG00000280008\n",
      "Missing ENSG00000280065\n",
      "Missing ENSG00000280074\n",
      "Missing ENSG00000280102\n",
      "Missing ENSG00000280111\n",
      "Missing ENSG00000280112\n",
      "Missing ENSG00000280133\n",
      "Missing ENSG00000280246\n",
      "Missing ENSG00000280253\n",
      "Missing ENSG00000280258\n",
      "Missing ENSG00000280315\n",
      "Missing ENSG00000280324\n",
      "Missing ENSG00000280328\n",
      "Missing ENSG00000280361\n",
      "Missing ENSG00000280370\n",
      "Missing ENSG00000280391\n",
      "Missing ENSG00000280413\n",
      "Missing ENSG00000280448\n",
      "Missing ENSG00000280456\n",
      "Missing ENSG00000280458\n",
      "Missing ENSG00000280459\n",
      "Missing ENSG00000280462\n",
      "Missing ENSG00000280464\n",
      "Missing ENSG00000280465\n",
      "Missing ENSG00000280468\n",
      "Missing ENSG00000280469\n",
      "Missing ENSG00000280472\n",
      "Missing ENSG00000280478\n",
      "Missing ENSG00000280480\n",
      "Missing ENSG00000280481\n",
      "Missing ENSG00000280483\n",
      "Missing ENSG00000280487\n",
      "Missing ENSG00000280488\n",
      "Missing ENSG00000280489\n",
      "Missing ENSG00000280492\n",
      "Missing ENSG00000280493\n",
      "Missing ENSG00000280499\n",
      "Missing ENSG00000280503\n",
      "Missing ENSG00000280507\n",
      "Missing ENSG00000280508\n",
      "Missing ENSG00000280509\n",
      "Missing ENSG00000280510\n",
      "Missing ENSG00000280517\n",
      "Missing ENSG00000280519\n",
      "Missing ENSG00000280522\n",
      "Missing ENSG00000280523\n",
      "Missing ENSG00000280524\n",
      "Missing ENSG00000280526\n",
      "Missing ENSG00000280527\n",
      "Missing ENSG00000280528\n",
      "Missing ENSG00000280533\n",
      "Missing ENSG00000280536\n",
      "Missing ENSG00000280539\n",
      "Missing ENSG00000280546\n",
      "Missing ENSG00000280547\n",
      "Missing ENSG00000280553\n",
      "Missing ENSG00000280558\n",
      "Missing ENSG00000280564\n",
      "Missing ENSG00000280565\n",
      "Missing ENSG00000280566\n",
      "Missing ENSG00000280567\n",
      "Missing ENSG00000280570\n",
      "Missing ENSG00000280573\n",
      "Missing ENSG00000280574\n",
      "Missing ENSG00000280576\n",
      "Missing ENSG00000280581\n",
      "Missing ENSG00000280583\n",
      "Missing ENSG00000280586\n",
      "Missing ENSG00000280588\n",
      "Missing ENSG00000280589\n",
      "Missing ENSG00000280595\n",
      "Missing ENSG00000280598\n",
      "Missing ENSG00000280599\n",
      "Missing ENSG00000280600\n",
      "Missing ENSG00000280602\n",
      "Missing ENSG00000280605\n",
      "Missing ENSG00000280608\n",
      "Missing ENSG00000280615\n",
      "Missing ENSG00000280622\n",
      "Missing ENSG00000280624\n",
      "Missing ENSG00000280625\n",
      "Missing ENSG00000280628\n",
      "Missing ENSG00000280630\n",
      "Missing ENSG00000280637\n",
      "Missing ENSG00000280642\n",
      "Missing ENSG00000280643\n",
      "Missing ENSG00000280644\n",
      "Missing ENSG00000280648\n",
      "Missing ENSG00000280653\n",
      "Missing ENSG00000280654\n",
      "Missing ENSG00000280656\n",
      "Missing ENSG00000280657\n",
      "Missing ENSG00000280659\n",
      "Missing ENSG00000280662\n",
      "Missing ENSG00000280667\n",
      "Missing ENSG00000280672\n",
      "Missing ENSG00000280673\n",
      "Missing ENSG00000280676\n",
      "Missing ENSG00000280678\n",
      "Missing ENSG00000280685\n",
      "Missing ENSG00000280687\n",
      "Missing ENSG00000280692\n",
      "Missing ENSG00000280694\n",
      "Missing ENSG00000280696\n",
      "Missing ENSG00000280697\n",
      "Missing ENSG00000280698\n",
      "Missing ENSG00000280699\n",
      "Missing ENSG00000280702\n",
      "Missing ENSG00000280713\n",
      "Missing ENSG00000280715\n",
      "Missing ENSG00000280718\n",
      "Missing ENSG00000280723\n",
      "Missing ENSG00000280727\n",
      "Missing ENSG00000280728\n",
      "Missing ENSG00000280729\n",
      "Missing ENSG00000280730\n",
      "Missing ENSG00000280731\n",
      "Missing ENSG00000280736\n",
      "Missing ENSG00000280738\n",
      "Missing ENSG00000280742\n",
      "Missing ENSG00000280743\n",
      "Missing ENSG00000280746\n",
      "Missing ENSG00000280747\n",
      "Missing ENSG00000280753\n",
      "Missing ENSG00000280756\n",
      "Missing ENSG00000280760\n",
      "Missing ENSG00000280761\n",
      "Missing ENSG00000280762\n",
      "Missing ENSG00000280763\n",
      "Missing ENSG00000280764\n",
      "Missing ENSG00000280766\n",
      "Missing ENSG00000280771\n",
      "Missing ENSG00000280772\n",
      "Missing ENSG00000280777\n",
      "Missing ENSG00000280779\n",
      "Missing ENSG00000280783\n",
      "Missing ENSG00000280785\n",
      "Missing ENSG00000280788\n",
      "Missing ENSG00000280790\n",
      "Missing ENSG00000280791\n",
      "Missing ENSG00000280793\n",
      "Missing ENSG00000280796\n",
      "Missing ENSG00000280801\n",
      "Missing ENSG00000280803\n",
      "Missing ENSG00000280804\n",
      "Missing ENSG00000280816\n",
      "Missing ENSG00000280818\n",
      "Missing ENSG00000280822\n",
      "Missing ENSG00000280823\n",
      "Missing ENSG00000280824\n",
      "Missing ENSG00000280826\n",
      "Missing ENSG00000280827\n",
      "Missing ENSG00000280829\n",
      "Missing ENSG00000280833\n",
      "Missing ENSG00000280834\n",
      "Missing ENSG00000280835\n",
      "Missing ENSG00000280842\n",
      "Missing ENSG00000280843\n",
      "Missing ENSG00000280845\n",
      "Missing ENSG00000280846\n",
      "Missing ENSG00000280847\n",
      "Missing ENSG00000280848\n",
      "Missing ENSG00000280849\n",
      "Missing ENSG00000280851\n",
      "Missing ENSG00000280854\n",
      "Missing ENSG00000280855\n",
      "Missing ENSG00000280856\n",
      "Missing ENSG00000280857\n",
      "Missing ENSG00000280859\n",
      "Missing ENSG00000280860\n",
      "Missing ENSG00000280861\n",
      "Missing ENSG00000280862\n",
      "Missing ENSG00000280863\n",
      "Missing ENSG00000280867\n",
      "Missing ENSG00000280869\n",
      "Missing ENSG00000280871\n",
      "Missing ENSG00000280874\n",
      "Missing ENSG00000280875\n",
      "Missing ENSG00000280884\n",
      "Missing ENSG00000280886\n",
      "Missing ENSG00000280887\n",
      "Missing ENSG00000280888\n",
      "Missing ENSG00000280896\n",
      "Missing ENSG00000280897\n",
      "Missing ENSG00000280900\n",
      "Missing ENSG00000280901\n",
      "Missing ENSG00000280902\n",
      "Missing ENSG00000280903\n",
      "Missing ENSG00000280910\n",
      "Missing ENSG00000280911\n",
      "Missing ENSG00000280914\n",
      "Missing ENSG00000280915\n",
      "Missing ENSG00000280918\n",
      "Missing ENSG00000280919\n",
      "Missing ENSG00000280921\n",
      "Missing ENSG00000280922\n",
      "Missing ENSG00000280925\n",
      "Missing ENSG00000280933\n",
      "Missing ENSG00000280934\n",
      "Missing ENSG00000280942\n",
      "Missing ENSG00000280945\n",
      "Missing ENSG00000280947\n",
      "Missing ENSG00000280950\n",
      "Missing ENSG00000280954\n",
      "Missing ENSG00000280955\n",
      "Missing ENSG00000280956\n",
      "Missing ENSG00000280957\n",
      "Missing ENSG00000280964\n",
      "Missing ENSG00000280965\n",
      "Missing ENSG00000280966\n",
      "Missing ENSG00000280970\n",
      "Missing ENSG00000280974\n",
      "Missing ENSG00000280976\n",
      "Missing ENSG00000280977\n",
      "Missing ENSG00000280979\n",
      "Missing ENSG00000280981\n",
      "Missing ENSG00000280985\n",
      "Missing ENSG00000280988\n",
      "Missing ENSG00000280992\n",
      "Missing ENSG00000280995\n",
      "Missing ENSG00000280998\n",
      "Missing ENSG00000280999\n",
      "Missing ENSG00000281001\n",
      "Missing ENSG00000281002\n",
      "Missing ENSG00000281003\n",
      "Missing ENSG00000281006\n",
      "Missing ENSG00000281007\n",
      "Missing ENSG00000281014\n",
      "Missing ENSG00000281017\n",
      "Missing ENSG00000281018\n",
      "Missing ENSG00000281023\n",
      "Missing ENSG00000281025\n",
      "Missing ENSG00000281027\n",
      "Missing ENSG00000281030\n",
      "Missing ENSG00000281033\n",
      "Missing ENSG00000281034\n",
      "Missing ENSG00000281035\n",
      "Missing ENSG00000281037\n",
      "Missing ENSG00000281041\n",
      "Missing ENSG00000281044\n",
      "Missing ENSG00000281045\n",
      "Missing ENSG00000281046\n",
      "Missing ENSG00000281047\n",
      "Missing ENSG00000281052\n",
      "Missing ENSG00000281057\n",
      "Missing ENSG00000281059\n",
      "Missing ENSG00000281061\n",
      "Missing ENSG00000281064\n",
      "Missing ENSG00000281066\n",
      "Missing ENSG00000281068\n",
      "Missing ENSG00000281069\n",
      "Missing ENSG00000281070\n",
      "Missing ENSG00000281075\n",
      "Missing ENSG00000281080\n",
      "Missing ENSG00000281085\n",
      "Missing ENSG00000281088\n",
      "Missing ENSG00000281090\n",
      "Missing ENSG00000281096\n",
      "Missing ENSG00000281098\n",
      "Missing ENSG00000281105\n",
      "Missing ENSG00000281108\n",
      "Missing ENSG00000281109\n",
      "Missing ENSG00000281110\n",
      "Missing ENSG00000281114\n",
      "Missing ENSG00000281115\n",
      "Missing ENSG00000281116\n",
      "Missing ENSG00000281117\n",
      "Missing ENSG00000281119\n",
      "Missing ENSG00000281124\n",
      "Missing ENSG00000281125\n",
      "Missing ENSG00000281129\n",
      "Missing ENSG00000281130\n",
      "Missing ENSG00000281134\n",
      "Missing ENSG00000281135\n",
      "Missing ENSG00000281136\n",
      "Missing ENSG00000281138\n",
      "Missing ENSG00000281140\n",
      "Missing ENSG00000281142\n",
      "Missing ENSG00000281143\n",
      "Missing ENSG00000281146\n",
      "Missing ENSG00000281147\n",
      "Missing ENSG00000281148\n",
      "Missing ENSG00000281152\n",
      "Missing ENSG00000281155\n",
      "Missing ENSG00000281158\n",
      "Missing ENSG00000281164\n",
      "Missing ENSG00000281169\n",
      "Missing ENSG00000281170\n",
      "Missing ENSG00000281171\n",
      "Missing ENSG00000281172\n",
      "Missing ENSG00000281177\n",
      "Missing ENSG00000281187\n",
      "Missing ENSG00000281190\n",
      "Missing ENSG00000281193\n",
      "Missing ENSG00000281194\n",
      "Missing ENSG00000281199\n",
      "Missing ENSG00000281200\n",
      "Missing ENSG00000281201\n",
      "Missing ENSG00000281203\n",
      "Missing ENSG00000281204\n",
      "Missing ENSG00000281206\n",
      "Missing ENSG00000281209\n",
      "Missing ENSG00000281212\n",
      "Missing ENSG00000281214\n",
      "Missing ENSG00000281215\n",
      "Missing ENSG00000281217\n",
      "Missing ENSG00000281218\n",
      "Missing ENSG00000281220\n",
      "Missing ENSG00000281222\n",
      "Missing ENSG00000281231\n",
      "Missing ENSG00000281235\n",
      "Missing ENSG00000281238\n",
      "Missing ENSG00000281239\n",
      "Missing ENSG00000281241\n",
      "Missing ENSG00000281242\n",
      "Missing ENSG00000281247\n",
      "Missing ENSG00000281253\n",
      "Missing ENSG00000281255\n",
      "Missing ENSG00000281257\n",
      "Missing ENSG00000281258\n",
      "Missing ENSG00000281259\n",
      "Missing ENSG00000281262\n",
      "Missing ENSG00000281271\n",
      "Missing ENSG00000281272\n",
      "Missing ENSG00000281275\n",
      "Missing ENSG00000281278\n",
      "Missing ENSG00000281280\n",
      "Missing ENSG00000281281\n",
      "Missing ENSG00000281285\n",
      "Missing ENSG00000281286\n",
      "Missing ENSG00000281288\n",
      "Missing ENSG00000281290\n",
      "Missing ENSG00000281291\n",
      "Missing ENSG00000281293\n",
      "Missing ENSG00000281294\n",
      "Missing ENSG00000281297\n",
      "Missing ENSG00000281298\n",
      "Missing ENSG00000281301\n",
      "Missing ENSG00000281302\n",
      "Missing ENSG00000281303\n",
      "Missing ENSG00000281304\n",
      "Missing ENSG00000281312\n",
      "Missing ENSG00000281314\n",
      "Missing ENSG00000281315\n",
      "Missing ENSG00000281318\n",
      "Missing ENSG00000281319\n",
      "Missing ENSG00000281322\n",
      "Missing ENSG00000281328\n",
      "Missing ENSG00000281329\n",
      "Missing ENSG00000281330\n",
      "Missing ENSG00000281337\n",
      "Missing ENSG00000281338\n",
      "Missing ENSG00000281339\n",
      "Missing ENSG00000281341\n",
      "Missing ENSG00000281342\n",
      "Missing ENSG00000281346\n",
      "Missing ENSG00000281350\n",
      "Missing ENSG00000281352\n",
      "Missing ENSG00000281353\n",
      "Missing ENSG00000281355\n",
      "Missing ENSG00000281366\n",
      "Missing ENSG00000281368\n",
      "Missing ENSG00000281372\n",
      "Missing ENSG00000281373\n",
      "Missing ENSG00000281375\n",
      "Missing ENSG00000281378\n",
      "Missing ENSG00000281381\n",
      "Missing ENSG00000281389\n",
      "Missing ENSG00000281390\n",
      "Missing ENSG00000281397\n",
      "Missing ENSG00000281405\n",
      "Missing ENSG00000281408\n",
      "Missing ENSG00000281411\n",
      "Missing ENSG00000281413\n",
      "Missing ENSG00000281414\n",
      "Missing ENSG00000281415\n",
      "Missing ENSG00000281419\n",
      "Missing ENSG00000281423\n",
      "Missing ENSG00000281428\n",
      "Missing ENSG00000281431\n",
      "Missing ENSG00000281433\n",
      "Missing ENSG00000281435\n",
      "Missing ENSG00000281437\n",
      "Missing ENSG00000281438\n",
      "Missing ENSG00000281439\n",
      "Missing ENSG00000281440\n",
      "Missing ENSG00000281441\n",
      "Missing ENSG00000281443\n",
      "Missing ENSG00000281444\n",
      "Missing ENSG00000281445\n",
      "Missing ENSG00000281447\n",
      "Missing ENSG00000281451\n",
      "Missing ENSG00000281452\n",
      "Missing ENSG00000281454\n",
      "Missing ENSG00000281457\n",
      "Missing ENSG00000281458\n",
      "Missing ENSG00000281461\n",
      "Missing ENSG00000281462\n",
      "Missing ENSG00000281463\n",
      "Missing ENSG00000281466\n",
      "Missing ENSG00000281467\n",
      "Missing ENSG00000281474\n",
      "Missing ENSG00000281475\n",
      "Missing ENSG00000281476\n",
      "Missing ENSG00000281479\n",
      "Missing ENSG00000281480\n",
      "Missing ENSG00000281483\n",
      "Missing ENSG00000281485\n",
      "Missing ENSG00000281488\n",
      "Missing ENSG00000281492\n",
      "Missing ENSG00000281497\n",
      "Missing ENSG00000281506\n",
      "Missing ENSG00000281512\n",
      "Missing ENSG00000281514\n",
      "Missing ENSG00000281515\n",
      "Missing ENSG00000281517\n",
      "Missing ENSG00000281522\n",
      "Missing ENSG00000281528\n",
      "Missing ENSG00000281531\n",
      "Missing ENSG00000281532\n",
      "Missing ENSG00000281534\n",
      "Missing ENSG00000281535\n",
      "Missing ENSG00000281539\n",
      "Missing ENSG00000281544\n",
      "Missing ENSG00000281545\n",
      "Missing ENSG00000281546\n",
      "Missing ENSG00000281549\n",
      "Missing ENSG00000281551\n",
      "Missing ENSG00000281554\n",
      "Missing ENSG00000281555\n",
      "Missing ENSG00000281558\n",
      "Missing ENSG00000281568\n",
      "Missing ENSG00000281570\n",
      "Missing ENSG00000281572\n",
      "Missing ENSG00000281574\n",
      "Missing ENSG00000281575\n",
      "Missing ENSG00000281576\n",
      "Missing ENSG00000281579\n",
      "Missing ENSG00000281581\n",
      "Missing ENSG00000281585\n",
      "Missing ENSG00000281587\n",
      "Missing ENSG00000281588\n",
      "Missing ENSG00000281589\n",
      "Missing ENSG00000281592\n",
      "Missing ENSG00000281594\n",
      "Missing ENSG00000281595\n",
      "Missing ENSG00000281596\n",
      "Missing ENSG00000281598\n",
      "Missing ENSG00000281599\n",
      "Missing ENSG00000281600\n",
      "Missing ENSG00000281605\n",
      "Missing ENSG00000281606\n",
      "Missing ENSG00000281611\n",
      "Missing ENSG00000281615\n",
      "Missing ENSG00000281619\n",
      "Missing ENSG00000281621\n",
      "Missing ENSG00000281628\n",
      "Missing ENSG00000281629\n",
      "Missing ENSG00000281631\n",
      "Missing ENSG00000281632\n",
      "Missing ENSG00000281634\n",
      "Missing ENSG00000281636\n",
      "Missing ENSG00000281637\n",
      "Missing ENSG00000281638\n",
      "Missing ENSG00000281639\n",
      "Missing ENSG00000281640\n",
      "Missing ENSG00000281643\n",
      "Missing ENSG00000281646\n",
      "Missing ENSG00000281647\n",
      "Missing ENSG00000281653\n",
      "Missing ENSG00000281662\n",
      "Missing ENSG00000281663\n",
      "Missing ENSG00000281666\n",
      "Missing ENSG00000281669\n",
      "Missing ENSG00000281672\n",
      "Missing ENSG00000281673\n",
      "Missing ENSG00000281675\n",
      "Missing ENSG00000281678\n",
      "Missing ENSG00000281681\n",
      "Missing ENSG00000281682\n",
      "Missing ENSG00000281684\n",
      "Missing ENSG00000281685\n",
      "Missing ENSG00000281686\n",
      "Missing ENSG00000281689\n",
      "Missing ENSG00000281694\n",
      "Missing ENSG00000281695\n",
      "Missing ENSG00000281698\n",
      "Missing ENSG00000281700\n",
      "Missing ENSG00000281701\n",
      "Missing ENSG00000281712\n",
      "Missing ENSG00000281713\n",
      "Missing ENSG00000281714\n",
      "Missing ENSG00000281715\n",
      "Missing ENSG00000281717\n",
      "Missing ENSG00000281719\n",
      "Missing ENSG00000281724\n",
      "Missing ENSG00000281726\n",
      "Missing ENSG00000281727\n",
      "Missing ENSG00000281729\n",
      "Missing ENSG00000281730\n",
      "Missing ENSG00000281735\n",
      "Missing ENSG00000281739\n",
      "Missing ENSG00000281744\n",
      "Missing ENSG00000281745\n",
      "Missing ENSG00000281746\n",
      "Missing ENSG00000281747\n",
      "Missing ENSG00000281749\n",
      "Missing ENSG00000281751\n",
      "Missing ENSG00000281752\n",
      "Missing ENSG00000281755\n",
      "Missing ENSG00000281765\n",
      "Missing ENSG00000281768\n",
      "Missing ENSG00000281774\n",
      "Missing ENSG00000281776\n",
      "Missing ENSG00000281779\n",
      "Missing ENSG00000281784\n",
      "Missing ENSG00000281786\n",
      "Missing ENSG00000281789\n",
      "Missing ENSG00000281790\n",
      "Missing ENSG00000281791\n",
      "Missing ENSG00000281795\n",
      "Missing ENSG00000281796\n",
      "Missing ENSG00000281802\n",
      "Missing ENSG00000281807\n",
      "Missing ENSG00000281812\n",
      "Missing ENSG00000281814\n",
      "Missing ENSG00000281816\n",
      "Missing ENSG00000281817\n",
      "Missing ENSG00000281823\n",
      "Missing ENSG00000281824\n",
      "Missing ENSG00000281826\n",
      "Missing ENSG00000281827\n",
      "Missing ENSG00000281834\n",
      "Missing ENSG00000281838\n",
      "Missing ENSG00000281843\n",
      "Missing ENSG00000281844\n",
      "Missing ENSG00000281845\n",
      "Missing ENSG00000281850\n",
      "Missing ENSG00000281851\n",
      "Missing ENSG00000281854\n",
      "Missing ENSG00000281864\n",
      "Missing ENSG00000281866\n",
      "Missing ENSG00000281867\n",
      "Missing ENSG00000281868\n",
      "Missing ENSG00000281869\n",
      "Missing ENSG00000281871\n",
      "Missing ENSG00000281872\n",
      "Missing ENSG00000281875\n",
      "Missing ENSG00000281882\n",
      "Missing ENSG00000281884\n",
      "Missing ENSG00000281888\n",
      "Missing ENSG00000281891\n",
      "Missing ENSG00000281892\n",
      "Missing ENSG00000281893\n",
      "Missing ENSG00000281895\n",
      "Missing ENSG00000281902\n",
      "Missing ENSG00000281905\n",
      "Missing ENSG00000281906\n",
      "Missing ENSG00000281918\n",
      "Missing ENSG00000281921\n",
      "Missing ENSGR0000002586\n",
      "Missing ENSGR0000124333\n",
      "Missing ENSGR0000124334\n",
      "Missing ENSGR0000167393\n",
      "Missing ENSGR0000168939\n",
      "Missing ENSGR0000169084\n",
      "Missing ENSGR0000169093\n",
      "Missing ENSGR0000169100\n",
      "Missing ENSGR0000178605\n",
      "Missing ENSGR0000182162\n",
      "Missing ENSGR0000182378\n",
      "Missing ENSGR0000182484\n",
      "Missing ENSGR0000185203\n",
      "Missing ENSGR0000185291\n",
      "Missing ENSGR0000185960\n",
      "Missing ENSGR0000196433\n",
      "Missing ENSGR0000197976\n",
      "Missing ENSGR0000198223\n",
      "Missing ENSGR0000205755\n",
      "Missing ENSGR0000214717\n",
      "Missing ENSGR0000223274\n",
      "Missing ENSGR0000223484\n",
      "Missing ENSGR0000223511\n",
      "Missing ENSGR0000223571\n",
      "Missing ENSGR0000223773\n",
      "Missing ENSGR0000225661\n",
      "Missing ENSGR0000226179\n",
      "Missing ENSGR0000227159\n",
      "Missing ENSGR0000228410\n",
      "Missing ENSGR0000228572\n",
      "Missing ENSGR0000229232\n",
      "Missing ENSGR0000230542\n",
      "Missing ENSGR0000234622\n",
      "Missing ENSGR0000234958\n",
      "Missing ENSGR0000236017\n",
      "Missing ENSGR0000236871\n",
      "Missing ENSGR0000237040\n",
      "Missing ENSGR0000237531\n",
      "Missing ENSGR0000237801\n",
      "Missing ENSGR0000263835\n",
      "Missing ENSGR0000263980\n",
      "Missing ENSGR0000264510\n",
      "Missing ENSGR0000264819\n",
      "Missing ENSGR0000265658\n",
      "Missing ENSGR0000270726\n",
      "Missing ENSGR0000275287\n",
      "Missing ENSGR0000276543\n",
      "Missing ENSGR0000277120\n",
      "Missing ENSGR0000280767\n",
      "Missing ENSGR0000281849\n"
     ]
    }
   ],
   "source": [
    "eprint(\"Importing all the correspondance tables\")\n",
    "# Tables 1 et 2: correspondance Ensembl <-> GeneNames (à refaire en SPARQL ?)\n",
    "\n",
    "filename = filenameCorrespondanceEnsemblGenenames\n",
    "\n",
    "EnsemblToGeneNames = dict()\n",
    "GeneNamesToEnsembl = dict()\n",
    "\n",
    "EnsemblToGeneNamesList = dict()\n",
    "GeneNamesToEnsemblList = dict()\n",
    "\n",
    "\n",
    "fd = open(filename,\"r\")\n",
    "\n",
    "lines = fd.readlines()\n",
    "for l in lines:\n",
    "    t=l.split(\";\")\n",
    "    t[2]=t[2].replace(\"\\n\",\"\")\n",
    "    EnsemblToGeneNames[t[0]]=t[2]\n",
    "    GeneNamesToEnsembl[t[2]]=t[0]\n",
    "    if (not EnsemblToGeneNamesList.get(t[0])):\n",
    "        EnsemblToGeneNamesList[t[0]]=list()\n",
    "    if (not GeneNamesToEnsemblList.get(t[2])):\n",
    "        GeneNamesToEnsemblList[t[2]]=list()\n",
    "    EnsemblToGeneNamesList[t[0]].append(t[2])\n",
    "    GeneNamesToEnsemblList[t[2]].append(t[0])\n",
    "    \n",
    "fd.close()\n",
    "\n",
    "# Missing genename associations\n",
    "GeneNamesToEnsembl[\"TGIF\"]=GeneNamesToEnsembl[\"TGIF1\"]\n",
    "GeneNamesToEnsembl[\"CART1\"]=GeneNamesToEnsembl[\"ALX1\"]\n",
    "EnsemblToGeneNames[\"ENSG00000122718\"]=\"OR2S2\"\n",
    "EnsemblToGeneNames[\"ENSG00000155640\"] = \"C10orf12\"\n",
    "GeneNamesToEnsembl[\"C10orf12\"]=\"ENSG00000155640\"\n",
    "EnsemblToGeneNames[\"ENSG00000168260\"] = \"C14orf183\"\n",
    "GeneNamesToEnsembl[\"C14orf183\"]=\"ENSG00000168260\"\n",
    "EnsemblToGeneNames[\"ENSG00000168787\"] = \"OR12D2\"\n",
    "EnsemblToGeneNames[\"ENSG00000170803\"] = \"OR2AG1\"\n",
    "EnsemblToGeneNames[\"ENSG00000171484\"] = \"OR1B1\"\n",
    "EnsemblToGeneNames[\"ENSG00000172381\"] = \"OR6Q1\"\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# Tables 3 et 4 : Correspondance (Enzymes,Transporters) <-> Ensembl\n",
    "\n",
    "filename = filenameCorrespondanceEnsemblEnzymes\n",
    "\n",
    "ECTCDBToEnsembl = dict()\n",
    "fd = open(filename,\"r\")\n",
    "\n",
    "lines = fd.readlines()\n",
    "for l in lines:\n",
    "    t=l.split(\";\")\n",
    "    L=t[1].replace(\"\\n\",\"\").split(\",\")\n",
    "    ECTCDBToEnsembl[t[0]]=list()\n",
    "    for g in L:\n",
    "        if (GeneNamesToEnsembl.get(g)):\n",
    "            ECTCDBToEnsembl[t[0]].append(GeneNamesToEnsembl.get(g))\n",
    "fd.close()\n",
    "\n",
    "\n",
    "filename = filenameCorrespondanceEnsemblTransporters\n",
    "\n",
    "fd = open(filename,\"r\")\n",
    "\n",
    "lines = fd.readlines()\n",
    "for l in lines:\n",
    "    t=l.split(\";\")\n",
    "    L=t[1].replace(\"\\n\",\"\").split(\",\")\n",
    "    ECTCDBToEnsembl[t[0]]=list()\n",
    "    for g in L:\n",
    "        if (GeneNamesToEnsembl.get(g)):\n",
    "            ECTCDBToEnsembl[t[0]].append(GeneNamesToEnsembl.get(g))\n",
    "fd.close()\n",
    "\n",
    "# Table 5 : Second filter, the genes present in the RRG must possess a RNASeq value. List of genes having a RNASeq value\n",
    "\n",
    "#Request = \"\"\"curl -o /dev/stdout \"http://bgee.org/?page=gene&gene_id=%%GENEID%%\" -q | grep \"Name\" | awk -F \"<td>|</td>|<th>|</th>\" '{for(i=1;i<NF;i++) {if ($i == \"Ensembl ID\") {printf(\"%s;;\",$(i+2));}; if ($i == \"Name\") {printf(\"%s\\\\n\",$(i+2));};          }    }'>>~/Downloads/missing.csv\"\"\"\n",
    "\n",
    "#fout=open(\"/Users/jeremiebourdon/Downloads/requete.sh\",\"w\")\n",
    "\n",
    "filename = Param_RNASeqFilename\n",
    "fd = open(filename,\"r\")\n",
    "l=fd.readline()\n",
    "RNASeqlistofgenes=list()\n",
    "#missing=\"\"\n",
    "for l in fd:\n",
    "    l=l.replace('\"','').replace('\\n','')\n",
    "    t=l.split(\",\")\n",
    "    geneid=t[0].split(\".\")[0]\n",
    "#    if not (EnsemblToGeneNames.get(geneid)):\n",
    "#        print(geneid)\n",
    "#        missing=(Request.replace(\"%%GENEID%%\",geneid))\n",
    "#        fout.write(missing+\"\\n\")\n",
    "    if (EnsemblToGeneNames.get(geneid)):\n",
    "#        RNASeqlistofgenes.append(GeneNamesToEnsembl[EnsemblToGeneNames[geneid]]) # a conversion Ensembl->Gene->Ensembl to get a unique Ensemble Id for a given gene name\n",
    "        RNASeqlistofgenes.append(geneid)\n",
    "    else:\n",
    "        eprint(\"Missing\",geneid)\n",
    "fd.close()\n",
    "\n",
    "#fout.close()\n",
    "\n",
    "\n",
    "# Table 6 : Drug targets\n",
    "fd=open(filenameKEGGDrugsGenes ,\"r\")\n",
    "\n",
    "lines = fd.readlines()\n",
    "\n",
    "Allmeds=dict()\n",
    "\n",
    "for l in lines:\n",
    "    t=l.split(\";\")\n",
    "    t[len(t)-1]=t[len(t)-1].replace(\"\\n\",\"\")\n",
    "    completeTargets=t[2].split('|')+t[3].replace(\"Enzyme:\",\"\").split(',')+t[4].split('|')\n",
    "    completeTargets=[x.strip().split(\" \")[0] for x in completeTargets if len(x)>2]\n",
    "    completeTargetsEnsembl=[GeneNamesToEnsembl.get(x) for x in completeTargets if (GeneNamesToEnsembl.get(x) != None)]\n",
    "    Allmeds[t[1].strip()]={'KeggId': t[0], 'TargetsGenes': completeTargets, 'TargetsGenesEnsembl': completeTargetsEnsembl }\n",
    "\n",
    "fd.close()\n",
    "\n",
    "# Table 6 : Mean Transcript length\n",
    "\n",
    "fd = open(filenameCorrespondanceEnsemblMeanLength,\"r\")\n",
    "\n",
    "lines=fd.readlines()\n",
    "\n",
    "AllGeneMeans = dict()\n",
    "for g in lines:\n",
    "    t=g.split(\";\")\n",
    "    t[1]=float(t[1].replace(\"\\n\",\"\"))\n",
    "    AllGeneMeans[t[0]]=t[1]\n",
    "    \n",
    "fd.close()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "for g in EnsemblToGeneNamesList:\n",
    "    EnsemblToGeneNamesList[g]=list(set(EnsemblToGeneNamesList.get(g)))\n",
    "    EnsemblToGeneNamesList.get(g).sort()\n",
    "    if (len(EnsemblToGeneNamesList.get(g))>1):\n",
    "        print(g,EnsemblToGeneNamesList.get(g).sort())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Construction de la liste des cibles"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Entering : Association RxN <-> Enzymes list construction\n"
     ]
    }
   ],
   "source": [
    "# étape 1 : construction du fichier d'association Reactions <-> EC numbers\n",
    "\n",
    "eprint(\"Entering : Association RxN <-> Enzymes list construction\")\n",
    "\n",
    "SBMLfile = Param_SBMLfile\n",
    "\n",
    "#SBMLfile = raw_input(\"Enter the SBML filename ? \")\n",
    "\n",
    "def ConstructAssociationList(SBMLfile, RXN_EC_file=\"/tmp/RXN_EC.txt\"):\n",
    "\n",
    "    import os\n",
    "    os.system(\"../scripts/Construct_List_RXN_Enzymes.sh \"+SBMLfile+\" \"+RXN_EC_file)\n",
    "    fd = open(RXN_EC_file,\"r\")\n",
    "    lines = fd.readlines()\n",
    "\n",
    "    genes = dict()\n",
    "    for l in lines:\n",
    "        t=l.split(\" \")\n",
    "        cum=''\n",
    "        for i in range(1,len(t)-1):\n",
    "            cum=cum+t[i]+\" \"\n",
    "        genes[t[0]]=cum\n",
    "    return genes\n",
    "\n",
    "genes=ConstructAssociationList(SBMLfile,RXN_EC_file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Entering : Single Reaction Analysis\n",
      "Execution time (Single Reaction Analysis) =  0.00010793498950079083 seconds\n"
     ]
    }
   ],
   "source": [
    "# Etape 2 : Single Reaction Analysis\n",
    "\n",
    "eprint(\"Entering : Single Reaction Analysis\")\n",
    "\n",
    "# perfoming a single_reaction_deletion analysis of the network\n",
    "\n",
    "model = cobra.io.read_sbml_model(SBMLfile)\n",
    "\n",
    "# the objective coefficient must be properly set\n",
    "model.reactions.CancerBiomass_OF.objective_coefficient=1\n",
    "\n",
    "# reference biomass growth\n",
    "fba=model.optimize().f\n",
    "\n",
    "# single reaction deletion analysis\n",
    "\n",
    "## DEBUG ONLY\n",
    "begin = timer()\n",
    "## END DEBUG ONLY\n",
    "if (Param_DoReconstruction):\n",
    "    rates, status = cobra.flux_analysis.single_reaction_deletion(model,solver=whichsolver)\n",
    "\n",
    "## DEBUG ONLY\n",
    "end = timer()\n",
    "eprint(\"Execution time (Single Reaction Analysis) = \",end-begin,\"seconds\")\n",
    "## END DEBUG ONLY"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# Etape 3 : Extraction des enzymes régulateurs (nécessite étape 1 et 2)\n",
    "\n",
    "def getRegulators(rates, percentage=1.1):\n",
    "    regulators = dict()\n",
    "\n",
    "    cpt=0\n",
    "    for r in rates:\n",
    "        if (rates.get(r)<percentage*fba):\n",
    "            if (genes.get(r)!=''):\n",
    "                # print(r,':',rates.get(r),\" \",genes.get(r))\n",
    "                cpt=cpt+1\n",
    "                l=genes.get(r).split(' ')\n",
    "                for i in l:\n",
    "                    if (i != ''):\n",
    "                        regulators[i]=1\n",
    "    return regulators\n",
    "\n",
    "if (Param_DoReconstruction):\n",
    "    regulators = getRegulators(rates, Param_Pourcentage)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# Etape 4 : récupération de la liste de gènes régulateurs\n",
    "def GeneListFromEC(ecnumber):\n",
    "    if (ecnumber.find('-')>0):\n",
    "        cum=list()\n",
    "        for i in range(1,500):\n",
    "            gl=ECTCDBToEnsembl.get(ecnumber.replace('-',str(i)))\n",
    "            if (gl != None and len(gl)>0):\n",
    "                cum+=gl\n",
    "        return cum\n",
    "    else:\n",
    "        gl=ECTCDBToEnsembl.get(ecnumber)\n",
    "        if (gl == None):\n",
    "            gl=list()\n",
    "        return gl\n",
    "\n",
    "if (Param_DoReconstruction):\n",
    "    TargetGenesEnsembl=list()\n",
    "    for ec in regulators:\n",
    "        TargetGenesEnsembl+=GeneListFromEC(ec.lower())\n",
    "    \n",
    "    TargetGenes = [EnsemblToGeneNames.get(x) for x in TargetGenesEnsembl]\n",
    "    \n",
    "    \n",
    "    TargetGenes=set(TargetGenes)\n",
    "    TargetGenes=list(TargetGenes)\n",
    "    \n",
    "    TargetGenesEnsembl=set(TargetGenesEnsembl)\n",
    "    TargetGenesEnsembl=list(TargetGenesEnsembl)\n",
    "    \n",
    "    eprint(\"Target gene list size =\",len(TargetGenes))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true
   },
   "source": [
    "Regulatory network reconstruction using PathwayCommons SPARQL endpoint"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# Algorithms\n",
    "\n",
    "def GetTFControllers(EnsemblId):\n",
    "    commonPCPrefixes = \"\"\"\n",
    "PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>\n",
    "PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>\n",
    "PREFIX owl: <http://www.w3.org/2002/07/owl#>\n",
    "PREFIX xsd: <http://www.w3.org/2001/XMLSchema#>\n",
    "PREFIX dc: <http://purl.org/dc/elements/1.1/>\n",
    "PREFIX dcterms: <http://purl.org/dc/terms/>\n",
    "PREFIX foaf: <http://xmlns.com/foaf/0.1/>\n",
    "PREFIX skos: <http://www.w3.org/2004/02/skos/core#>\n",
    "PREFIX bp3: <http://www.biopax.org/release/biopax-level3.owl#>\n",
    "PREFIX taxon: <http://identifiers.org/taxonomy/>\n",
    "PREFIX reactome: <http://identifiers.org/reactome/>\n",
    "PREFIX release: <http://www.reactome.org/biopax/49/48887#>\n",
    "\n",
    "PREFIX up: <http://purl.uniprot.org/core/> \n",
    "PREFIX uniprot: <http://purl.uniprot.org/uniprot/>\n",
    "PREFIX bp: <http://www.biopax.org/release/biopax-level3.owl#>\n",
    "PREFIX chebi: <http://purl.obolibrary.org/obo/CHEBI_>\n",
    "PREFIX obo2: <http://purl.obolibrary.org/obo#>\n",
    "\"\"\"\n",
    "\n",
    "    Request = \"\"\"\n",
    "SELECT ?tempReac ?type ?controlledName ?controllerName ?source WHERE{ \n",
    "    FILTER( (?controlledName = 'Transcription of %%GENETARGETNAME%%'^^<http://www.w3.org/2001/XMLSchema#string>)\n",
    "        and (?controllerName != '%%GENETARGETNAME%%')\n",
    "        and (?source != 'mirtarbase') ) .\n",
    "    ?tempReac a bp:TemplateReactionRegulation .\n",
    "    ?tempReac bp:displayName ?reacName ; \n",
    "        bp:controlled ?controlled ; \n",
    "        bp:controller ?controller ; \n",
    "        bp:controlType ?type ; \n",
    "        bp:dataSource ?source .\n",
    "    ?controlled bp:displayName ?controlledName .\n",
    "    ?controller bp:displayName ?controllerName . }\n",
    "GROUP BY ?controlledName ?controllerName\n",
    "\"\"\"\n",
    "    RequestToBePerformed=commonPCPrefixes+Request.replace(\"%%GENETARGETNAME%%\",EnsemblId)\n",
    "    sparql = SPARQLWrapper(PC_Endpoint)\n",
    "    sparql.setQuery(RequestToBePerformed)\n",
    "    sparql.setReturnFormat(JSON)\n",
    "    query = sparql.query().convert()\n",
    "    results=list()\n",
    "    listofgenes=list()\n",
    "    ListBindings=query.get('results').get('bindings')\n",
    "    for l in ListBindings:\n",
    "        controllerName = (l.get('controllerName').get('value'))\n",
    "        controlType= (l.get('type').get('value'))\n",
    "        rule = controllerName+\"_\"+controlType+\"_\"+EnsemblId\n",
    "        if (rule not in results):\n",
    "            results.append(rule)\n",
    "        if (controllerName not in listofgenes):\n",
    "            listofgenes.append(controllerName)\n",
    "    return results,listofgenes\n",
    "\n",
    "def getParentGraph(l,limit=30):\n",
    "    i=0\n",
    "    allgenes=dict()\n",
    "    for g in l:\n",
    "        allgenes[g]=1\n",
    "    listofrelations=list()\n",
    "    unmapped=list()\n",
    "    while (i<len(l) and i<limit):\n",
    "        gene=l[i]\n",
    "        relations,newgenes = GetTFControllers(gene)\n",
    "        if (len(relations) == 0):\n",
    "            unmapped.append(gene)\n",
    "        listofrelations+=relations\n",
    "        for g in newgenes:\n",
    "            if (allgenes.get(g) != 1): # (g not in l):\n",
    "                l.append(g)\n",
    "                allgenes[g]=1\n",
    "        i+=1\n",
    "    return listofrelations,l,unmapped\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Entering : RRG reconstruction\n"
     ]
    }
   ],
   "source": [
    "eprint(\"Entering : RRG reconstruction\")\n",
    "\n",
    "\n",
    "## DEBUG ONLY\n",
    "begin = timer()\n",
    "## END DEBUG ONLY\n",
    "\n",
    "if (Param_DoReconstruction):\n",
    "    TargetGenesName = TargetGenes #[EnsemblToGeneNames.get(x) for x in TargetGenes if (len(x) > 2)]\n",
    "    \n",
    "    a,b,u=getParentGraph(TargetGenesName,50000)\n",
    "    \n",
    "    ## DEBUG ONLY\n",
    "    end = timer()\n",
    "    eprint(\"Execution time for RRG reconstruction = \",end-begin,\"seconds\")\n",
    "    ## END DEBUG ONLY\n",
    "    \n",
    "    eprint('Reconstructed graph contains',len(b),'genes,',len(a),'interactions,','with',len(u),'unmapped genes.')\n",
    "    \n",
    "    \n",
    "    # write the file for compatibility purpose\n",
    "    \n",
    "    fd = open(Param_RRGFilename,\"w\")\n",
    "    for r in a:\n",
    "        relation=r.split(\"_\")\n",
    "        if (relation[1] != \"ACTIVATION\"):\n",
    "            eprint(relation)\n",
    "        fd.write(relation[0]+\";\"+relation[2]+\"\\n\")\n",
    "    fd.close()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Entering : Graph Filtering\n",
      "Warning :  ENSG00000180318 or ENSG00000115919 isn't in RNASeq\n",
      "Warning :  ENSG00000177426 or ENSG00000104313 isn't in RNASeq\n",
      "Warning :  ENSG00000180318 or ENSG00000104313 isn't in RNASeq\n",
      "Warning :  ENSG00000180318 or ENSG00000089250 isn't in RNASeq\n",
      "Warning :  ENSG00000177426 or ENSG00000064655 isn't in RNASeq\n",
      "Warning :  ENSG00000180318 or ENSG00000008086 isn't in RNASeq\n",
      "Warning :  ENSG00000177426 or ENSG00000161896 isn't in RNASeq\n",
      "Warning :  ENSG00000180318 or ENSG00000077264 isn't in RNASeq\n",
      "Warning :  ENSG00000177426 or ENSG00000133216 isn't in RNASeq\n",
      "Warning :  ENSG00000177426 or ENSG00000116539 isn't in RNASeq\n",
      "Warning :  ENSG00000177426 or ENSG00000155897 isn't in RNASeq\n",
      "Warning :  ENSG00000180318 or ENSG00000081237 isn't in RNASeq\n",
      "Warning :  ENSG00000177426 or ENSG00000146872 isn't in RNASeq\n",
      "Warning :  ENSG00000180318 or ENSG00000175175 isn't in RNASeq\n",
      "Warning :  ENSG00000177426 or ENSG00000115170 isn't in RNASeq\n",
      "Warning :  ENSG00000180318 or ENSG00000275176 isn't in RNASeq\n",
      "Warning :  ENSG00000177426 or ENSG00000115687 isn't in RNASeq\n",
      "Warning :  ENSG00000177426 or ENSG00000087095 isn't in RNASeq\n",
      "Warning :  ENSG00000180318 or ENSG00000233276 isn't in RNASeq\n",
      "Warning :  ENSG00000180318 or ENSG00000135333 isn't in RNASeq\n",
      "Warning :  ENSG00000177426 or ENSG00000143322 isn't in RNASeq\n",
      "Warning :  ENSG00000177426 or ENSG00000154358 isn't in RNASeq\n",
      "Warning :  ENSG00000177426 or ENSG00000109107 isn't in RNASeq\n",
      "Warning :  ENSG00000180318 or ENSG00000110422 isn't in RNASeq\n",
      "Warning :  ENSG00000177426 or ENSG00000044524 isn't in RNASeq\n",
      "Warning :  ENSG00000177426 or ENSG00000104375 isn't in RNASeq\n",
      "Warning :  ENSG00000180318 or ENSG00000010810 isn't in RNASeq\n",
      "Warning :  ENSG00000180318 or ENSG00000095596 isn't in RNASeq\n",
      "Warning :  ENSG00000177426 or ENSG00000095596 isn't in RNASeq\n",
      "Warning :  ENSG00000177426 or ENSG00000173153 isn't in RNASeq\n",
      "Warning :  ENSG00000177426 or ENSG00000106546 isn't in RNASeq\n",
      "Warning :  ENSG00000177426 or ENSG00000100644 isn't in RNASeq\n",
      "Warning :  ENSG00000180318 or ENSG00000156925 isn't in RNASeq\n",
      "Warning :  ENSG00000177426 or ENSG00000163064 isn't in RNASeq\n",
      "Warning :  ENSG00000180318 or ENSG00000163064 isn't in RNASeq\n",
      "Warning :  ENSG00000180318 or ENSG00000143190 isn't in RNASeq\n",
      "Warning :  ENSG00000177426 or ENSG00000113580 isn't in RNASeq\n",
      "Warning :  ENSG00000177426 or ENSG00000118689 isn't in RNASeq\n",
      "Warning :  ENSG00000180318 or ENSG00000185630 isn't in RNASeq\n",
      "Warning :  ENSG00000180318 or ENSG00000179456 isn't in RNASeq\n",
      "Warning :  ENSG00000177426 or ENSG00000125952 isn't in RNASeq\n",
      "Warning :  ENSG00000180318 or ENSG00000168267 isn't in RNASeq\n",
      "Warning :  ENSG00000180318 or ENSG00000134532 isn't in RNASeq\n",
      "Warning :  ENSG00000180318 or ENSG00000106004 isn't in RNASeq\n",
      "Warning :  ENSG00000177426 or ENSG00000168269 isn't in RNASeq\n",
      "Warning :  ENSG00000180318 or ENSG00000105997 isn't in RNASeq\n",
      "Remove  116 genes\n"
     ]
    }
   ],
   "source": [
    "# Construction and filter\n",
    "eprint(\"Entering : Graph Filtering\")\n",
    "\n",
    "RNASeqList=dict()\n",
    "for g in RNASeqlistofgenes:\n",
    "    RNASeqList[EnsemblToGeneNames[g]]=1\n",
    "\n",
    "\n",
    "GenesToBeRemoved=list()\n",
    "\n",
    "filename = Param_RRGFilename\n",
    "\n",
    "genesingraph = dict()\n",
    "listofgenes = list()\n",
    "listofinteractions = list()\n",
    "listofinteractionsNamed = list()\n",
    "\n",
    "fd = open(filename,\"r\")\n",
    "\n",
    "lines = fd.readlines()\n",
    "geneid=0\n",
    "for l in lines:\n",
    "    t=l.split(\";\")\n",
    "    t[1]=t[1].replace(\"\\n\",\"\")\n",
    "#    if (GeneNamesToEnsembl[t[0]] in RNASeqlistofgenes and GeneNamesToEnsembl[t[1]] in RNASeqlistofgenes):\n",
    "    if (RNASeqList.get(t[0]) and RNASeqList.get(t[1])):\n",
    "        if (not (GeneNamesToEnsembl[t[0]] in GenesToBeRemoved or GeneNamesToEnsembl[t[1]] in GenesToBeRemoved)):\n",
    "            if (listofgenes.count(GeneNamesToEnsembl[t[0]]) == 0):\n",
    "                listofgenes.append(GeneNamesToEnsembl[t[0]])\n",
    "                genesingraph[GeneNamesToEnsembl[t[0]]]=len(listofgenes)-1\n",
    "            if (listofgenes.count(GeneNamesToEnsembl[t[1]]) == 0):\n",
    "                listofgenes.append(GeneNamesToEnsembl[t[1]])\n",
    "                genesingraph[GeneNamesToEnsembl[t[1]]]=len(listofgenes)-1\n",
    "            listofinteractions.append((genesingraph[GeneNamesToEnsembl[t[0]]],genesingraph[GeneNamesToEnsembl[t[1]]]))\n",
    "            listofinteractionsNamed.append((GeneNamesToEnsembl[t[0]],GeneNamesToEnsembl[t[1]]))\n",
    "    else:\n",
    "        eprint(\"Warning : \",GeneNamesToEnsembl[t[0]],\"or\",GeneNamesToEnsembl[t[1]],\"isn't in RNASeq\",)\n",
    "\n",
    "fd.close()\n",
    "\n",
    "\n",
    "\n",
    "# It is required to remove cycles and loops in the graph.\n",
    "# Here is a greedy method\n",
    "\n",
    "listofinteractionsNamedClean = list()\n",
    "\n",
    "RRG = nx.DiGraph()\n",
    "for t in listofinteractionsNamed:\n",
    "#    if True:\n",
    "    if not((RRG.has_node(t[0]) and RRG.has_node(t[1]) and (RRG.has_edge(t[1],t[0]) or nx.has_path(RRG,t[1],t[0])))):\n",
    "        RRG.add_edge(t[0],t[1])\n",
    "        listofinteractionsNamedClean.append((t[0],t[1]))\n",
    "\n",
    "globalmax=100\n",
    "nbRemovedGenes=0\n",
    "\n",
    "while (globalmax>Param_MaxIndegreeInRRG):\n",
    "    T=RRG.in_degree()\n",
    "    Tout=RRG.out_degree()\n",
    "    vmax=-1\n",
    "    globalmax=-1\n",
    "    keymax = \"\"\n",
    "    for g in T:\n",
    "        v=T.get(g)\n",
    "        if (v>vmax and Tout.get(g)>0):\n",
    "            vmax=v\n",
    "            keymax = g\n",
    "        if (v>globalmax):\n",
    "            globalmax=v\n",
    "    #print(\"Remove gene \",keymax, \"having an in degree equal to\",vmax)\n",
    "    GenesToBeRemoved.append(keymax)\n",
    "    nbRemovedGenes+=1\n",
    "    RRG.remove_node(keymax)\n",
    "eprint(\"Remove \",nbRemovedGenes,\"genes\")\n",
    "\n",
    "filename = Param_RRGFilename\n",
    "\n",
    "genesingraph = dict()\n",
    "listofgenes = list()\n",
    "listofinteractions = list()\n",
    "listofinteractionsNamed = list()\n",
    "\n",
    "fd = open(filename,\"r\")\n",
    "\n",
    "lines = fd.readlines()\n",
    "geneid=0\n",
    "for l in lines:\n",
    "    t=l.split(\";\")\n",
    "    t[1]=t[1].replace(\"\\n\",\"\")\n",
    "#    if (GeneNamesToEnsembl[t[0]] in RNASeqlistofgenes and GeneNamesToEnsembl[t[1]] in RNASeqlistofgenes):\n",
    "#    if (RNASeqList.get(GeneNamesToEnsembl[t[0]]) and RNASeqList.get(GeneNamesToEnsembl[t[1]])):\n",
    "    if (RNASeqList.get(t[0]) and RNASeqList.get(t[1])):\n",
    "        if (not (GeneNamesToEnsembl[t[0]] in GenesToBeRemoved or GeneNamesToEnsembl[t[1]] in GenesToBeRemoved)):\n",
    "            if (listofgenes.count(GeneNamesToEnsembl[t[0]]) == 0):\n",
    "                listofgenes.append(GeneNamesToEnsembl[t[0]])\n",
    "                genesingraph[GeneNamesToEnsembl[t[0]]]=len(listofgenes)-1\n",
    "            if (listofgenes.count(GeneNamesToEnsembl[t[1]]) == 0):\n",
    "                listofgenes.append(GeneNamesToEnsembl[t[1]])\n",
    "                genesingraph[GeneNamesToEnsembl[t[1]]]=len(listofgenes)-1\n",
    "            listofinteractions.append((genesingraph[GeneNamesToEnsembl[t[0]]],genesingraph[GeneNamesToEnsembl[t[1]]]))\n",
    "            listofinteractionsNamed.append((GeneNamesToEnsembl[t[0]],GeneNamesToEnsembl[t[1]]))\n",
    "\n",
    "fd.close()\n",
    "\n",
    "# It is required to remove to cycles and loops in the graph.\n",
    "# Here is a greedy method\n",
    "\n",
    "listofinteractionsNamedClean = list()\n",
    "\n",
    "RRG = nx.DiGraph()\n",
    "for t in listofinteractionsNamed:\n",
    "    if not((RRG.has_node(t[0]) and RRG.has_node(t[1]) and (RRG.has_edge(t[1],t[0]) or nx.has_path(RRG,t[1],t[0])))):\n",
    "        RRG.add_edge(t[0],t[1])\n",
    "        listofinteractionsNamedClean.append((t[0],t[1]))\n",
    "#    else:\n",
    "#        print(\"Boucle si ajout de %s -> %s\" % (t[0],t[1]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Entering : Bayesian model Skeleton Construction\n",
      "Execution time (Bayesian model Skeleton Construction) =  0.06754369499685708 seconds\n"
     ]
    }
   ],
   "source": [
    "# Building the Bayesian model (from listofinteractionsNamedClean)\n",
    "\n",
    "eprint(\"Entering : Bayesian model Skeleton Construction\")\n",
    "\n",
    "## DEBUG ONLY\n",
    "begin = timer()\n",
    "## END DEBUG ONLY\n",
    "\n",
    "\n",
    "\n",
    "model = BayesianModel(listofinteractionsNamedClean)\n",
    "\n",
    "\n",
    "\n",
    "## DEBUG ONLY\n",
    "end = timer()\n",
    "eprint(\"Execution time (Bayesian model Skeleton Construction) = \",end-begin,\"seconds\")\n",
    "## END DEBUG ONLY"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# basic optimization (use of a hash table to know if a gene is in the RRG)\n",
    "GeneDict=dict()\n",
    "\n",
    "for g in listofgenes:\n",
    "    GeneDict[g]=1\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "RNASeq Data integration"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Entering : RNASeq data discretization\n"
     ]
    },
    {
     "ename": "ValueError",
     "evalue": "Shape of passed values is (589, 250), indices imply (584, 250)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[0;32m/Users/jeremiebourdon/anaconda/lib/python3.5/site-packages/pandas/core/internals.py\u001b[0m in \u001b[0;36mcreate_block_manager_from_blocks\u001b[0;34m(blocks, axes)\u001b[0m\n\u001b[1;32m   3983\u001b[0m                 blocks = [make_block(values=blocks[0],\n\u001b[0;32m-> 3984\u001b[0;31m                                      placement=slice(0, len(axes[0])))]\n\u001b[0m\u001b[1;32m   3985\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Users/jeremiebourdon/anaconda/lib/python3.5/site-packages/pandas/core/internals.py\u001b[0m in \u001b[0;36mmake_block\u001b[0;34m(values, placement, klass, ndim, dtype, fastpath)\u001b[0m\n\u001b[1;32m   2517\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 2518\u001b[0;31m     \u001b[0;32mreturn\u001b[0m \u001b[0mklass\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mvalues\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mndim\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mndim\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mfastpath\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mfastpath\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mplacement\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mplacement\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   2519\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Users/jeremiebourdon/anaconda/lib/python3.5/site-packages/pandas/core/internals.py\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, values, placement, ndim, fastpath)\u001b[0m\n\u001b[1;32m     89\u001b[0m                              'implies %d' % (len(self.values),\n\u001b[0;32m---> 90\u001b[0;31m                                              len(self.mgr_locs)))\n\u001b[0m\u001b[1;32m     91\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mValueError\u001b[0m: Wrong number of items passed 589, placement implies 584",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-43-dc2de07731ec>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m     58\u001b[0m \u001b[0mfd\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mclose\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     59\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 60\u001b[0;31m \u001b[0mdata\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpd\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mDataFrame\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mraw_data\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcolumns\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0msortedlistofgenes\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     61\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     62\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Users/jeremiebourdon/anaconda/lib/python3.5/site-packages/pandas/core/frame.py\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, data, index, columns, dtype, copy)\u001b[0m\n\u001b[1;32m    253\u001b[0m             \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    254\u001b[0m                 mgr = self._init_ndarray(data, index, columns, dtype=dtype,\n\u001b[0;32m--> 255\u001b[0;31m                                          copy=copy)\n\u001b[0m\u001b[1;32m    256\u001b[0m         \u001b[0;32melif\u001b[0m \u001b[0misinstance\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdata\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mlist\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtypes\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mGeneratorType\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    257\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0misinstance\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdata\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtypes\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mGeneratorType\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Users/jeremiebourdon/anaconda/lib/python3.5/site-packages/pandas/core/frame.py\u001b[0m in \u001b[0;36m_init_ndarray\u001b[0;34m(self, values, index, columns, dtype, copy)\u001b[0m\n\u001b[1;32m    430\u001b[0m             \u001b[0mvalues\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0m_possibly_infer_to_datetimelike\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mvalues\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    431\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 432\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mcreate_block_manager_from_blocks\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mvalues\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0mcolumns\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mindex\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    433\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    434\u001b[0m     \u001b[0;34m@\u001b[0m\u001b[0mproperty\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Users/jeremiebourdon/anaconda/lib/python3.5/site-packages/pandas/core/internals.py\u001b[0m in \u001b[0;36mcreate_block_manager_from_blocks\u001b[0;34m(blocks, axes)\u001b[0m\n\u001b[1;32m   3991\u001b[0m         \u001b[0mblocks\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0mgetattr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mb\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'values'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mb\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mb\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mblocks\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3992\u001b[0m         \u001b[0mtot_items\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0msum\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mb\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mshape\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mb\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mblocks\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 3993\u001b[0;31m         \u001b[0mconstruction_error\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtot_items\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mblocks\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mshape\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maxes\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0me\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   3994\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3995\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/Users/jeremiebourdon/anaconda/lib/python3.5/site-packages/pandas/core/internals.py\u001b[0m in \u001b[0;36mconstruction_error\u001b[0;34m(tot_items, block_shape, axes, e)\u001b[0m\n\u001b[1;32m   3968\u001b[0m         \u001b[0;32mraise\u001b[0m \u001b[0mValueError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"Empty data passed with indices specified.\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3969\u001b[0m     raise ValueError(\"Shape of passed values is {0}, indices imply {1}\".format(\n\u001b[0;32m-> 3970\u001b[0;31m         passed, implied))\n\u001b[0m\u001b[1;32m   3971\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3972\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mValueError\u001b[0m: Shape of passed values is (589, 250), indices imply (584, 250)"
     ]
    }
   ],
   "source": [
    "# Reading the datafile (RNA-seq normalized with DESeq2)\n",
    "\n",
    "eprint(\"Entering : RNASeq data discretization\")\n",
    "\n",
    "\n",
    "## DEBUG ONLY\n",
    "begin = timer()\n",
    "## END DEBUG ONLY\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "filename = Param_RNASeqFilename\n",
    "\n",
    "data = dict()\n",
    "\n",
    "TumoralConditions=list()\n",
    "NormalConditions=list()\n",
    "\n",
    "fd = open(filename,\"r\")\n",
    "\n",
    "l=fd.readline()\n",
    "l=l.replace('\"','').replace('\\n','')\n",
    "t=l.split(\",\")\n",
    "\n",
    "for i in range(1,len(t)):\n",
    "    if (t[i][0:2] == \"CT\"):\n",
    "        TumoralConditions.append(i)\n",
    "    if (t[i][0:2] == \"CN\"):\n",
    "        NormalConditions.append(i)\n",
    "\n",
    "# à résoudre: il y a des gènes dans le graphe qui ne sont pas dans les données... Arrrgh\n",
    "\n",
    "raw_data=np.random.randint(low=0, high=1, size=(len(TumoralConditions),len(listofgenes)))\n",
    "\n",
    "sortedlistofgenes=list()\n",
    "# allgenes=list()\n",
    "\n",
    "for l in fd:\n",
    "    l=l.replace('\"','').replace('\\n','')\n",
    "    t=l.split(\",\")\n",
    "    NormalValues = [log(0.0001+float(t[x])) for x in NormalConditions]\n",
    "#    NormalValues=np.array(len(NormalConditions))\n",
    "#    for j in range(len(NormalConditions)):\n",
    "#        NormalValues[j]=t[NormalConditions[j]]\n",
    "    NormalMean=np.mean(NormalValues)\n",
    "    NormalStd=np.std(NormalValues)\n",
    "    geneid=t[0].split(\".\")[0]\n",
    "    BinarizedValues=[Binarized(float(t[x]),NormalMean,NormalStd,geneid) for x in TumoralConditions]\n",
    "    #if (GeneDict.get(geneid) == 1):\n",
    "    #        print(geneid,sum(BinarizedValues))\n",
    "    #if (geneid in listofgenes):\n",
    "    if (GeneDict.get(geneid) == 1):\n",
    "        raw_data[:,len(sortedlistofgenes)]=BinarizedValues\n",
    "        sortedlistofgenes.append(geneid)\n",
    "    #print(t[0].split(\".\")[0],BinarizedValues)\n",
    "\n",
    "fd.close()\n",
    "\n",
    "data = pd.DataFrame(raw_data, columns=sortedlistofgenes)\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "## DEBUG ONLY\n",
    "end = timer()\n",
    "eprint(\"Execution time (RNASeq data discretization) = \",end-begin,\"seconds\")\n",
    "## END DEBUG ONLY\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# Fitting the BN with data\n",
    "\n",
    "eprint(\"Entering : Bayesian Network fitting with RNASeq data\")\n",
    "\n",
    "## DEBUG ONLY\n",
    "begin = timer()\n",
    "## END DEBUG ONLY\n",
    "\n",
    "model.fit(data, estimator=BayesianEstimator, prior_type='K2')\n",
    "\n",
    "## DEBUG ONLY\n",
    "end = timer()\n",
    "eprint(\"Execution time (Bayesian Network fitting with RNASeq data) = \",end-begin,\"seconds\")\n",
    "## END DEBUG ONLY\n",
    "\n",
    "# saving the Bayesian Network for further use\n",
    "writer = BIFWriter(model)\n",
    "writer.write_bif(filename=BIF_file)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Querying the Bayesian Network"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def getEvidence(medname):\n",
    "    listgenes = Allmeds.get(medname).get(\"TargetsGenesEnsembl\")\n",
    "    evidences = [State(x, 0) for x in listgenes if x in RRG.nodes()]\n",
    "    return evidences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# An alternative way is to perform a Bayesian Sampling of the BN\n",
    "\n",
    "eprint(\"Entering : Querying the Bayesian Network\")\n",
    "\n",
    "\n",
    "## DEBUG ONLY\n",
    "begin = timer()\n",
    "## END DEBUG ONLY\n",
    "\n",
    "inference = BayesianModelSampling(model)\n",
    "## DEBUG ONLY\n",
    "end = timer()\n",
    "eprint(\"Execution time, first part of Querying the Bayesian Network = \",end-begin,\"seconds\")\n",
    "## END DEBUG ONLY\n",
    "\n",
    "\n",
    "Number_of_samples = Param_NumberOfSample\n",
    "\n",
    "# NB evidences should be read somewhere\n",
    "evidences = getEvidence(Param_Drug)\n",
    "\n",
    "\n",
    "## DEBUG ONLY\n",
    "begin = timer()\n",
    "## END DEBUG ONLY\n",
    "\n",
    "samples=inference.likelihood_weighted_sample(evidences, Number_of_samples)\n",
    "#samples=inference.rejection_sample(evidences, Number_of_samples)\n",
    "\n",
    "\n",
    "## DEBUG ONLY\n",
    "end = timer()\n",
    "eprint(\"Execution time, second part (Querying the Bayesian Network) = \",end-begin,\"seconds\")\n",
    "## END DEBUG ONLY"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# get all the marginal probabilities for that evidence\n",
    "allMarginalProb=dict()\n",
    "for g in sortedlistofgenes:\n",
    "    #print(g,\" p0=\",samples[g].eq(0).sum()/Number_of_samples,\" p1=\",samples[g].eq(1).sum()/Number_of_samples,\" p2=\",samples[g].eq(2).sum()/Number_of_samples)\n",
    "    allMarginalProb[g]=[samples[g].eq(0).sum()/Number_of_samples,samples[g].eq(1).sum()/Number_of_samples,samples[g].eq(2).sum()/Number_of_samples]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# Correspondance Reaction <-> Enzymes\n",
    "filename = RXN_EC_file\n",
    "\n",
    "ReactionsToEnsembl = dict()\n",
    "\n",
    "fd = open(filename,\"r\")\n",
    "\n",
    "lines = fd.readlines()\n",
    "for l in lines:\n",
    "    t=l.split(\" \")\n",
    "    RXN = t[0]\n",
    "    t[len(t)-1]=t[len(t)-1].replace(\"\\n\",\"\")\n",
    "    ReactionsToEnsembl[RXN]=list()\n",
    "    for i in range(1,len(t)):\n",
    "        val = ECTCDBToEnsembl.get(t[i].lower())\n",
    "        if (val != None):\n",
    "            ReactionsToEnsembl[RXN]+=val\n",
    "fd.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "def ProbaReaction(L):\n",
    "    global allMarginalProb\n",
    "    maximum=0\n",
    "    minimum=1000    \n",
    "    for l in L:\n",
    "        if allMarginalProb.get(l):\n",
    "            minimum=min([minimum,1-allMarginalProb.get(l)[0]])\n",
    "            maximum=max([maximum,1-allMarginalProb.get(l)[0]])\n",
    "    if minimum==1000:\n",
    "        return \"1;1\"\n",
    "    else:\n",
    "        return str(minimum)+\";\"+str(maximum)\n",
    "\n",
    "\n",
    "\n",
    "# Output the reactions probabilities\n",
    "\n",
    "output=open(Param_ProbaReactionsFilename,\"w\")\n",
    "\n",
    "for r in ReactionsToEnsembl:\n",
    "    if (len(ReactionsToEnsembl.get(r))>0):\n",
    "        output.write(str(r)+\";\"+str(ProbaReaction(ReactionsToEnsembl.get(r)))+\"\\n\")\n",
    "output.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Dernière partie, calcul de la nouvelle biomasse"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# To be very efficient, the long formal expressions must be reconstructed in a clever way\n",
    "\n",
    "def ExpressionFromList_rec(L,a,b):\n",
    "    #print(L,a,b)\n",
    "    if (b == a):\n",
    "        #print(\"Adding\",0)\n",
    "        return 0\n",
    "    else:\n",
    "        if (b-a == 1):\n",
    "            #print(\"Adding L[\"+str(a)+\"]\",L[a])\n",
    "            return L[a]\n",
    "        else:\n",
    "            milieu = int((a+b)/2)\n",
    "            return ExpressionFromList_rec(L,a,milieu)+ExpressionFromList_rec(L,milieu,b)\n",
    "\n",
    "def ExpressionFromList(L):\n",
    "    return ExpressionFromList_rec(L,0,len(L))\n",
    "\n",
    "\n",
    "def CobraModelToPROMOptLangModel(model,Prob,lambd=-1.0,usedsolver=\"cglpk\"):\n",
    "    epsilon = 1e-10\n",
    "    myoptmodel = optlang.Model(name = \"test\")\n",
    "    # first of all, performs a FVA analysis\n",
    "    eprint(\"Performing a FVA analysis\")\n",
    "    currenttime=timer()\n",
    "    fva0=flux_variability_analysis(model,fraction_of_optimum=0.001,solver=usedsolver)\n",
    "    fva=flux_variability_analysis(model,fraction_of_optimum=0.99,solver=usedsolver)\n",
    "    eprint(\"Finishing the FVA analysis in\",timer()-currenttime,\"seconds\")\n",
    "\n",
    "\n",
    "    obj=0\n",
    "    ListObjective=list()\n",
    "\n",
    "    Allvariables = list()\n",
    "    Alphavariables = list()\n",
    "    Betavariables = list()\n",
    "\n",
    "    AdditionalConstraints = list()\n",
    "    i=0\n",
    "    currenttime=timer()\n",
    "    \n",
    "    \n",
    "    for r in model.reactions:\n",
    "        if (i % 100 == 0):\n",
    "            eprint(\"Creating\",i,\"variables in\",timer()-currenttime,\"seconds\")\n",
    "            currenttime=timer()\n",
    "        i+=1\n",
    "\n",
    "        #v=optlang.Variable(r.id,lb=r.lower_bound,ub=r.upper_bound)\n",
    "        v=optlang.Variable(r.id,lb=fva0.get(r.id).get('minimum')-epsilon,ub=fva0.get(r.id).get('maximum')+epsilon)     \n",
    "        # v=optlang.Variable(r.id,lb=-2000,ub=2000)\n",
    "        Allvariables.append(v)\n",
    "\n",
    "        # alpha_i and beta_i\n",
    "        pi = Prob.get(r.id) # now a vector of [min,max] probabilities\n",
    "        if not pi:\n",
    "            pi = [1,1]\n",
    "        if (pi[0]<0.99):        \n",
    "            alpha=optlang.Variable('alpha_'+r.id,lb=0,ub=100000)\n",
    "            beta=optlang.Variable('beta_'+r.id,lb=0,ub=100000)\n",
    "            Alphavariables.append(alpha)\n",
    "            Betavariables.append(beta)\n",
    "            # contributions of alpha and beta to the objective = lambda * (alpha+beta) with lambda = -1\n",
    "            #obj = obj + lambd * alpha + lambd * beta\n",
    "            ListObjective.append(lambd * alpha)\n",
    "            ListObjective.append(lambd * beta)\n",
    "            \n",
    "            AdditionalConstraints.append(optlang.Constraint(v+alpha,lb=pi[1]*fva.get(r.id).get('minimum'),ub=2000))\n",
    "            AdditionalConstraints.append(optlang.Constraint(v-beta,lb=-2000,ub=pi[0]*fva.get(r.id).get('maximum')))\n",
    "\n",
    "        if r.objective_coefficient != 0:\n",
    "            # obj = obj + r.objective_coefficient * v\n",
    "            ListObjective.append(r.objective_coefficient * v)\n",
    "\n",
    "    \n",
    "    currenttime=timer()\n",
    "    eprint(\"Reconstructing Objective\")\n",
    "#    for o in ListObjective:\n",
    "#        obj=obj+o\n",
    "    obj = ExpressionFromList(ListObjective)\n",
    "    eprint(\"Reconstructing Objective in\",timer()-currenttime,\"seconds\")\n",
    "    \n",
    "    AllConstraints = list()\n",
    "    i=0\n",
    "    currenttime=timer()\n",
    "    for m in model.metabolites:\n",
    "        if (i % 100 == 0):\n",
    "            eprint(\"Creating\",i,\"constraints in\",timer()-currenttime,\"seconds\")\n",
    "            currenttime=timer()\n",
    "        i+=1\n",
    "        C=list()\n",
    "        for r in m.reactions:\n",
    "            C.append(r.get_coefficient(m)*Allvariables[model.reactions.index(r)])\n",
    "        AllConstraints.append(optlang.Constraint(ExpressionFromList(C),lb=0,ub=0))\n",
    "\n",
    "    eprint(\"Adding Allvariables\")\n",
    "    myoptmodel.add(Allvariables)\n",
    "    eprint(\"Adding Alphavariables\")\n",
    "    myoptmodel.add(Alphavariables)\n",
    "    eprint(\"Adding Betavariables\")\n",
    "    myoptmodel.add(Betavariables)\n",
    "    eprint(\"Adding AllConstraints\")\n",
    "    myoptmodel.add(AllConstraints)\n",
    "    eprint(\"Adding AdditionalConstraints\")\n",
    "    myoptmodel.add(AdditionalConstraints)\n",
    "    myoptmodel.objective = optlang.Objective(obj, direction=\"max\")\n",
    "    return myoptmodel\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# on recharge les probabilités des réactions pour des problèmes de compatibilités\n",
    "\n",
    "filename = Param_ProbaReactionsFilename\n",
    "fd=open(filename,\"r\")\n",
    "\n",
    "ProbaReaction=dict()\n",
    "\n",
    "lines = fd.readlines()\n",
    "\n",
    "for l in lines:\n",
    "    t=l.split(\";\")\n",
    "    t[1]=t[1].replace(\"\\n\",\"\")\n",
    "    ProbaReaction[t[0]]=[float(t[1]),float(t[2])]\n",
    "\n",
    "fd.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# on recharge le modèle cobra\n",
    "mcancer = cobra.io.read_sbml_model(Param_SBMLfile)\n",
    "mcancer.reactions.CancerBiomass_OF.objective_coefficient = 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Performing a FVA analysis\n",
      "Finishing the FVA analysis in 76.13515910299611 seconds\n",
      "Creating 0 variables in 6.417001713998616e-06 seconds\n",
      "Creating 100 variables in 0.030046086001675576 seconds\n",
      "Creating 200 variables in 0.0203411279944703 seconds\n",
      "Creating 300 variables in 0.010559315996943042 seconds\n",
      "Creating 400 variables in 0.012518830000772141 seconds\n",
      "Creating 500 variables in 0.024819788988679647 seconds\n",
      "Creating 600 variables in 0.01967674499610439 seconds\n",
      "Creating 700 variables in 0.015266921996953897 seconds\n",
      "Creating 800 variables in 0.014040042995475233 seconds\n",
      "Creating 900 variables in 0.011274874996161088 seconds\n",
      "Creating 1000 variables in 0.08095710199268069 seconds\n",
      "Creating 1100 variables in 0.1867264239990618 seconds\n",
      "Creating 1200 variables in 0.005916485999478027 seconds\n",
      "Creating 1300 variables in 0.01320988500083331 seconds\n",
      "Creating 1400 variables in 0.019400478995521553 seconds\n",
      "Creating 1500 variables in 0.04817806000937708 seconds\n",
      "Creating 1600 variables in 0.0031690759933553636 seconds\n",
      "Creating 1700 variables in 0.01826351099589374 seconds\n",
      "Creating 1800 variables in 0.028528194990940392 seconds\n",
      "Creating 1900 variables in 0.011076031005359255 seconds\n",
      "Creating 2000 variables in 0.009096594003494829 seconds\n",
      "Creating 2100 variables in 0.006940927007235587 seconds\n",
      "Creating 2200 variables in 0.046684043991263025 seconds\n",
      "Creating 2300 variables in 0.09375382200232707 seconds\n",
      "Creating 2400 variables in 0.015189646001090296 seconds\n",
      "Creating 2500 variables in 0.019523658993421122 seconds\n",
      "Creating 2600 variables in 0.01469865300168749 seconds\n",
      "Creating 2700 variables in 0.013686356993275695 seconds\n",
      "Creating 2800 variables in 0.003939035988878459 seconds\n",
      "Creating 2900 variables in 0.007446809991961345 seconds\n",
      "Creating 3000 variables in 0.009837412988417782 seconds\n",
      "Creating 3100 variables in 0.015926922991639003 seconds\n",
      "Creating 3200 variables in 0.02326843500486575 seconds\n",
      "Creating 3300 variables in 0.025805875993683003 seconds\n",
      "Creating 3400 variables in 0.015393075009342283 seconds\n",
      "Creating 3500 variables in 0.01176388499152381 seconds\n",
      "Creating 3600 variables in 0.037220494996290654 seconds\n",
      "Creating 3700 variables in 0.02658603100280743 seconds\n",
      "Creating 3800 variables in 0.004345388995716348 seconds\n",
      "Creating 3900 variables in 0.019068911002250388 seconds\n",
      "Creating 4000 variables in 0.005567524000070989 seconds\n",
      "Creating 4100 variables in 0.006077224999899045 seconds\n",
      "Creating 4200 variables in 0.0054606829944532365 seconds\n",
      "Creating 4300 variables in 0.003749022987904027 seconds\n",
      "Reconstructing Objective\n",
      "Reconstructing Objective in 3.3227467329998035 seconds\n",
      "Creating 0 constraints in 2.6439956855028868e-06 seconds\n",
      "Creating 100 constraints in 0.10732738900696859 seconds\n",
      "Creating 200 constraints in 0.0949693699949421 seconds\n",
      "Creating 300 constraints in 0.12793338700430468 seconds\n",
      "Creating 400 constraints in 0.09718989100656472 seconds\n",
      "Creating 500 constraints in 0.09863689199846704 seconds\n",
      "Creating 600 constraints in 0.08612152400019113 seconds\n",
      "Creating 700 constraints in 0.07760658500774298 seconds\n",
      "Creating 800 constraints in 0.06442352299927734 seconds\n",
      "Creating 900 constraints in 0.06379382099839859 seconds\n",
      "Creating 1000 constraints in 0.05630172100791242 seconds\n",
      "Creating 1100 constraints in 0.06606741400901228 seconds\n",
      "Creating 1200 constraints in 0.0722536449902691 seconds\n",
      "Creating 1300 constraints in 0.11045158799970523 seconds\n",
      "Creating 1400 constraints in 0.4210732690116856 seconds\n",
      "Creating 1500 constraints in 0.6858793729916215 seconds\n",
      "Creating 1600 constraints in 0.22451510399696417 seconds\n",
      "Creating 1700 constraints in 0.06666701199719682 seconds\n",
      "Creating 1800 constraints in 0.1220121329970425 seconds\n",
      "Creating 1900 constraints in 1.1117296960001113 seconds\n",
      "Creating 2000 constraints in 0.0965191940049408 seconds\n",
      "Creating 2100 constraints in 0.13515087599807885 seconds\n",
      "Creating 2200 constraints in 0.2087875110009918 seconds\n",
      "Creating 2300 constraints in 0.2040262890077429 seconds\n",
      "Creating 2400 constraints in 2.585072854999453 seconds\n",
      "Creating 2500 constraints in 0.08199741899443325 seconds\n",
      "Creating 2600 constraints in 0.09777572000166401 seconds\n",
      "Creating 2700 constraints in 0.07595156699244399 seconds\n",
      "Creating 2800 constraints in 0.37060559599194676 seconds\n",
      "Creating 2900 constraints in 0.15597878900007345 seconds\n",
      "Creating 3000 constraints in 1.5615662539930781 seconds\n",
      "Creating 3100 constraints in 0.49823358899448067 seconds\n",
      "Creating 3200 constraints in 0.23690037199412473 seconds\n",
      "Creating 3300 constraints in 0.5729069119988708 seconds\n",
      "Creating 3400 constraints in 0.1870405809895601 seconds\n",
      "Creating 3500 constraints in 0.19951709899760317 seconds\n",
      "Creating 3600 constraints in 0.08836949399847072 seconds\n",
      "Creating 3700 constraints in 0.3215925690019503 seconds\n",
      "Creating 3800 constraints in 0.016125961992656812 seconds\n",
      "Creating 3900 constraints in 0.004587500006891787 seconds\n",
      "Creating 4000 constraints in 0.004864007991272956 seconds\n",
      "Creating 4100 constraints in 0.004492406995268539 seconds\n",
      "Creating 4200 constraints in 0.004547219010419212 seconds\n",
      "Creating 4300 constraints in 0.004970887006493285 seconds\n",
      "Creating 4400 constraints in 0.005295830997056328 seconds\n",
      "Creating 4500 constraints in 0.005173491997993551 seconds\n",
      "Creating 4600 constraints in 0.004780839997692965 seconds\n",
      "Creating 4700 constraints in 0.0055556860024807975 seconds\n",
      "Creating 4800 constraints in 0.004543547998764552 seconds\n",
      "Creating 4900 constraints in 0.005146640003658831 seconds\n",
      "Creating 5000 constraints in 0.00675992199103348 seconds\n",
      "Creating 5100 constraints in 0.006119033991126344 seconds\n",
      "Creating 5200 constraints in 0.005560357996728271 seconds\n",
      "Creating 5300 constraints in 0.004985684005077928 seconds\n",
      "Creating 5400 constraints in 0.004643287000362761 seconds\n",
      "Adding Allvariables\n",
      "Adding Alphavariables\n",
      "Adding Betavariables\n",
      "Adding AllConstraints\n",
      "Adding AdditionalConstraints\n",
      "Execution time (PROM Model creation) = 96.81688027000928 seconds\n"
     ]
    }
   ],
   "source": [
    "# Creation du modèle PROM sous optlang\n",
    "begin = timer()\n",
    "mProm=CobraModelToPROMOptLangModel(mcancer,ProbaReaction,usedsolver=whichsolver)\n",
    "end = timer()\n",
    "\n",
    "eprint(\"Execution time (PROM Model creation) =\",end-begin,\"seconds\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Optimize a model with 7134 rows, 7736 columns and 24296 nonzeros\n",
      "Coefficient statistics:\n",
      "  Matrix range     [1e-04, 8e+04]\n",
      "  Objective range  [1e+00, 1e+00]\n",
      "  Bounds range     [2e-12, 1e+05]\n",
      "  RHS range        [7e-15, 2e+03]\n",
      "Presolve removed 6249 rows and 5733 columns\n",
      "Presolve time: 0.02s\n",
      "Presolved: 885 rows, 2003 columns, 8390 nonzeros\n",
      "\n",
      "Iteration    Objective       Primal Inf.    Dual Inf.      Time\n",
      "       0    5.0591486e+03   3.797524e+05   0.000000e+00      0s\n",
      "Extra 281 simplex iterations after uncrush\n",
      "    1110   -2.0133030e+01   0.000000e+00   0.000000e+00      0s\n",
      "\n",
      "Solved in 1110 iterations and 0.09 seconds\n",
      "Optimal objective -2.013303011e+01\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Gurobi 0.0891809849999845 seconds, biomass= 0.9792230891827144\n",
      "glpk 1.0316394029941875 seconds, biommass= 0.9792231215846577\n"
     ]
    }
   ],
   "source": [
    "# Compares glpk and gurobi\n",
    "if debugon:\n",
    "    import gurobipy\n",
    "\n",
    "    lp = mProm.to_lp()\n",
    "    lpfilename = Param_SBMLfileBN.replace(\".xml\",\".lp\")\n",
    "    fd=open(lpfilename,\"w\")\n",
    "    fd.write(lp)\n",
    "    fd.close()\n",
    "\n",
    "    mPromGurobi = gurobipy.read(lpfilename)\n",
    "    \n",
    "    begin = timer()\n",
    "    res=mPromGurobi.optimize()\n",
    "    eprint(\"Gurobi\",timer()-begin,\"seconds, biomass=\",mPromGurobi.getVarByName(\"CancerBiomass_OF\").X)\n",
    "\n",
    "\n",
    "    begin = timer()\n",
    "    res=mProm.optimize()\n",
    "    eprint(\"glpk\",timer()-begin,\"seconds, biommass=\",mProm.variables.CancerBiomass_OF.primal)    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Execution time (PROM Optimization) = 0.0036909530026605353 seconds\n"
     ]
    }
   ],
   "source": [
    "# Optimisation\n",
    "begin = timer()\n",
    "mProm.optimize()\n",
    "end = timer()\n",
    "\n",
    "eprint(\"Execution time (PROM Optimization) =\",end-begin,\"seconds\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Biomass =  0.9792231215846577\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Objective value =  -20.133030226594947\n"
     ]
    }
   ],
   "source": [
    "# Results\n",
    "eprint(\"Objective value = \",mProm.objective.value)\n",
    "print(\"Biomass = \",mProm.variables.CancerBiomass_OF.primal)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Execution completed in 16114.383364281995 seconds\n"
     ]
    }
   ],
   "source": [
    "endtotal = timer()\n",
    "eprint(\"Execution completed in\",endtotal-begintotal,\"seconds\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "print(GeneNamesToEnsembl[EnsemblToGeneNames['ENSG00000204231']])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "anaconda-cloud": {},
  "kernelspec": {
   "display_name": "Python [conda root]",
   "language": "python",
   "name": "conda-root-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
